// @react/react-spectrum/NumberInput/js/NumberInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _autobindDecorator = _interopRequireDefault(require("autobind-decorator"));

var _Button = _interopRequireDefault(require("../../Button"));

var _events = require("../../utils/events");

var _ChevronDownSmall = _interopRequireDefault(require("../../Icon/core/ChevronDownSmall"));

var _ChevronUpSmall = _interopRequireDefault(require("../../Icon/core/ChevronUpSmall"));

var _number = require("../../utils/number");

var _classnames = _interopRequireDefault(require("classnames"));

var _convertUnsafeMethod = _interopRequireDefault(require("../../utils/convertUnsafeMethod"));

var _createId = _interopRequireDefault(require("../../utils/createId"));

var _filterDOMProps = _interopRequireDefault(require("../../utils/filterDOMProps"));

var _InputGroup = _interopRequireDefault(require("../../InputGroup"));

var _LiveRegionAnnouncer = _interopRequireDefault(require("../../utils/LiveRegionAnnouncer"));

var _intl = require("../../utils/intl");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireWildcard(require("react"));

var _reactDom = _interopRequireDefault(require("react-dom"));

var _Textfield = _interopRequireDefault(require("../../Textfield"));

var _class, _class2, _temp;

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; if (obj != null) { var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var intlMessages = {
  "cs-CZ": _interopRequireDefault(require("./../intl/cs-CZ.json"))["default"],
  "da-DK": _interopRequireDefault(require("./../intl/da-DK.json"))["default"],
  "de-DE": _interopRequireDefault(require("./../intl/de-DE.json"))["default"],
  "en-US": _interopRequireDefault(require("./../intl/en-US.json"))["default"],
  "es-ES": _interopRequireDefault(require("./../intl/es-ES.json"))["default"],
  "fi-FI": _interopRequireDefault(require("./../intl/fi-FI.json"))["default"],
  "fr-FR": _interopRequireDefault(require("./../intl/fr-FR.json"))["default"],
  "hu-HU": _interopRequireDefault(require("./../intl/hu-HU.json"))["default"],
  "it-IT": _interopRequireDefault(require("./../intl/it-IT.json"))["default"],
  "ja-JP": _interopRequireDefault(require("./../intl/ja-JP.json"))["default"],
  "ko-KR": _interopRequireDefault(require("./../intl/ko-KR.json"))["default"],
  "nb-NO": _interopRequireDefault(require("./../intl/nb-NO.json"))["default"],
  "nl-NL": _interopRequireDefault(require("./../intl/nl-NL.json"))["default"],
  "pl-PL": _interopRequireDefault(require("./../intl/pl-PL.json"))["default"],
  "pt-BR": _interopRequireDefault(require("./../intl/pt-BR.json"))["default"],
  "ru-RU": _interopRequireDefault(require("./../intl/ru-RU.json"))["default"],
  "sv-SE": _interopRequireDefault(require("./../intl/sv-SE.json"))["default"],
  "tr-TR": _interopRequireDefault(require("./../intl/tr-TR.json"))["default"],
  "uk-UA": _interopRequireDefault(require("./../intl/uk-UA.json"))["default"],
  "zh-CN": _interopRequireDefault(require("./../intl/zh-CN.json"))["default"],
  "zh-TW": _interopRequireDefault(require("./../intl/zh-TW.json"))["default"]
};

if (process.env.SCALE_MEDIUM && process.env.SCALE_LARGE) {
  require("../../spectrum-css/stepper/index.css");

  require("../../spectrum-css/stepper/index-diff.css");
} else if (process.env.SCALE_LARGE) {
  require("../../spectrum-css/stepper/index-lg.css");
} else {
  require("../../spectrum-css/stepper/index.css");
}

if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_LIGHT) require("../../spectrum-css/stepper/multiStops/light.css");
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_DARK) require("../../spectrum-css/stepper/multiStops/dark.css");
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_LIGHTEST) require("../../spectrum-css/stepper/multiStops/lightest.css");
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_DARKEST) require("../../spectrum-css/stepper/multiStops/darkest.css");
var formatMessage = (0, _intl.messageFormatter)(intlMessages);

var NumberInput = (0, _convertUnsafeMethod["default"])(_class = (0, _autobindDecorator["default"])(_class = (_temp = _class2 =
/*#__PURE__*/
function (_Component) {
  _inherits(NumberInput, _Component);

  function NumberInput(props) {
    var _this;

    _classCallCheck(this, NumberInput);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(NumberInput).call(this, props));
    var id = props.id,
        value = props.value,
        defaultValue = props.defaultValue;
    _this.state = {
      focused: false,
      valueInvalid: false,
      inputId: id || (0, _createId["default"])(),
      value: _this.getValue(value, defaultValue)
    };
    return _this;
  }

  _createClass(NumberInput, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      if (this.textfield) {
        _reactDom["default"].findDOMNode(this.textfield).addEventListener('wheel', this.handleInputScrollWheel, {
          passive: false,
          bubbles: false,
          cancelable: true
        });
      }
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      if (this.textfield) {
        _reactDom["default"].findDOMNode(this.textfield).removeEventListener('wheel', this.handleInputScrollWheel);
      }
    }
    /**
     * @private
     * Takes to value and returns the first defined value or equal to zero
     */

  }, {
    key: "getValue",
    value: function getValue(value, defaultValue) {
      if (value || value === 0) {
        return value;
      } else if (defaultValue || defaultValue === 0) {
        return defaultValue;
      }

      return '';
    }
  }, {
    key: "UNSAFE_componentWillReceiveProps",
    value: function UNSAFE_componentWillReceiveProps(nextProps) {
      if ('value' in nextProps) {
        this.setState({
          value: this.getValue(nextProps.value),
          valueInvalid: this.isInputValueInvalid(nextProps.value)
        });
      }

      if ('defaultValue' in nextProps) {
        this.setState({
          valueInvalid: this.isInputValueInvalid(nextProps.defaultValue)
        });
      }
    }
  }, {
    key: "onMouseDown",
    value: function onMouseDown(e) {
      e.preventDefault(); // Don't shift focus to textfield when increment/decrement button is pressed on a mobile device.

      !this.flagTouchStart && this.textfield.focus();
    }
    /**
    * Keep track of when increment/decrement button is pressed on a mobile device.
    */

  }, {
    key: "onTouchStart",
    value: function onTouchStart() {
      this.flagTouchStart = true;
    }
  }, {
    key: "handleDecrementButtonClick",
    value: function handleDecrementButtonClick(e) {
      e.preventDefault();
      this.decrementValue();
    }
  }, {
    key: "handleIncrementButtonClick",
    value: function handleIncrementButtonClick(e) {
      e.preventDefault();
      this.incrementValue();
    }
  }, {
    key: "handleFocus",
    value: function handleFocus() {
      this.setState({
        focused: true
      });
    }
  }, {
    key: "handleBlur",
    value: function handleBlur() {
      this.flagTouchStart = false;
      this.setState({
        focused: false
      });

      _LiveRegionAnnouncer["default"].clearMessage();
    }
  }, {
    key: "handleInputKeyDown",
    value: function handleInputKeyDown(e) {
      switch (e.keyCode) {
        case 38: // up arrow

        case 33:
          // page up
          e.preventDefault();
          this.incrementValue();
          break;

        case 40: // down arrow

        case 34:
          // page down
          e.preventDefault();
          this.decrementValue();
          break;

        case 35:
          // end
          e.preventDefault();
          this.decrementToMinValue();
          break;

        case 36:
          // home
          e.preventDefault();
          this.incrementToMaxValue();
          break;

        default: // do nothing

      }
    }
  }, {
    key: "handleInputScrollWheel",
    value: function handleInputScrollWheel(e) {
      e.preventDefault(); // If the input isn't supposed to receive input, do nothing.

      if (this.isInactive()) {
        return;
      }

      if (e.deltaY < 0) {
        this.incrementValue();
      } else {
        this.decrementValue();
      }
    }
  }, {
    key: "handleInputChange",
    value: function handleInputChange(value, e) {
      e.stopPropagation();
      var onChange = this.props.onChange;
      var valueAsNumber = value === '' ? null : +value;
      var numeric = !isNaN(valueAsNumber); // They may be starting to type a negative number, we don't want to broadcast this to
      // the onChange handler, but we do want to update the value state.

      var resemblesNumber = numeric || value === '-' || value === ''; // Only dispatch a change event if it's an actual number

      if (numeric || value === '') {
        onChange(valueAsNumber);
      }

      if (resemblesNumber) {
        this.setState({
          value: value,
          valueInvalid: this.isInputValueInvalid(value)
        });
      }
    }
    /**
     * @private
     * Returns true if the Textfield cannot receive any input.
     */

  }, {
    key: "isInactive",
    value: function isInactive() {
      var _this$props = this.props,
          disabled = _this$props.disabled,
          readOnly = _this$props.readOnly;
      var focused = this.state.focused;
      return disabled || readOnly || !focused;
    }
    /**
     * @private
     * Adds step number to the value number so long as it stays within min/max (if they are defined).
     * If value is undefined, sets it to the step.
     */

  }, {
    key: "incrementValue",
    value: function incrementValue() {
      var _this$props2 = this.props,
          min = _this$props2.min,
          max = _this$props2.max,
          step = _this$props2.step;
      var value = this.state.value;

      if (isNaN(step)) {
        step = 1;
      }

      var newValue = +value;

      if (isNaN(newValue)) {
        newValue = max != null ? Math.min(step, max) : step;
      } else {
        newValue = (0, _number.clamp)((0, _number.handleDecimalOperation)('+', newValue, step), min, max);
      }

      this.triggerChange(newValue);
    }
    /**
     * @private
     * If max is defined, sets value to the max value.
     */

  }, {
    key: "incrementToMaxValue",
    value: function incrementToMaxValue() {
      var max = this.props.max; // If the input isn't supposed to receive input, do nothing.

      if (this.isInactive()) {
        return;
      }

      if (max != null) {
        this.triggerChange(max);
      }
    }
    /**
     * @private
     * Subtracts step number from the value number so long as it stays within min/max (if they are
     * defined). If value is undefined, sets it to the step.
     */

  }, {
    key: "decrementValue",
    value: function decrementValue() {
      var _this$props3 = this.props,
          min = _this$props3.min,
          max = _this$props3.max,
          step = _this$props3.step;
      var value = this.state.value;

      if (isNaN(step)) {
        step = 1;
      }

      var newValue = +value;

      if (isNaN(newValue)) {
        newValue = min != null ? Math.max(-step, min) : -step;
      } else {
        newValue = (0, _number.clamp)((0, _number.handleDecimalOperation)('-', newValue, step), min, max);
      }

      this.triggerChange(newValue);
    }
    /**
     * @private
     * If min is defined, sets value to the max value.
     */

  }, {
    key: "decrementToMinValue",
    value: function decrementToMinValue() {
      var min = this.props.min; // If the input isn't supposed to receive input, do nothing.

      if (this.isInactive()) {
        return;
      }

      if (min != null) {
        this.triggerChange(min);
      }
    }
    /**
     * @private
     * Updates state to the new value and notifies the onChange handler.
     */

  }, {
    key: "triggerChange",
    value: function triggerChange(newValue) {
      var onChange = this.props.onChange;
      var value = this.state.value; // Only trigger change event and setState if the value changed

      if (value !== newValue) {
        // Announce new value using a live region
        _LiveRegionAnnouncer["default"].announceAssertive(newValue.toString());

        this.setState({
          value: newValue,
          valueInvalid: this.isInputValueInvalid(newValue)
        });
        onChange(newValue);
      }
    }
    /**
     * @private
     * If true, the input's value is currently invalid.
     */

  }, {
    key: "isInputValueInvalid",
    value: function isInputValueInvalid(value) {
      var _this$props4 = this.props,
          max = _this$props4.max,
          min = _this$props4.min;
      return value !== '' && isNaN(+value) || max !== null && value > max || min !== null && value < min;
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props5 = this.props,
          defaultValue = _this$props5.defaultValue,
          _this$props5$placehol = _this$props5.placeholder,
          placeholder = _this$props5$placehol === void 0 ? formatMessage('Enter a number') : _this$props5$placehol,
          min = _this$props5.min,
          max = _this$props5.max,
          step = _this$props5.step,
          _this$props5$decremen = _this$props5.decrementTitle,
          decrementTitle = _this$props5$decremen === void 0 ? formatMessage('Decrement') : _this$props5$decremen,
          _this$props5$incremen = _this$props5.incrementTitle,
          incrementTitle = _this$props5$incremen === void 0 ? formatMessage('Increment') : _this$props5$incremen,
          invalid = _this$props5.invalid,
          disabled = _this$props5.disabled,
          quiet = _this$props5.quiet,
          className = _this$props5.className,
          readOnly = _this$props5.readOnly,
          ariaLabel = _this$props5['aria-label'],
          ariaLabelledby = _this$props5['aria-labelledby'],
          otherProps = _objectWithoutProperties(_this$props5, ["defaultValue", "placeholder", "min", "max", "step", "decrementTitle", "incrementTitle", "invalid", "disabled", "quiet", "className", "readOnly", "aria-label", "aria-labelledby"]);

      var _this$state = this.state,
          focused = _this$state.focused,
          value = _this$state.value,
          valueInvalid = _this$state.valueInvalid,
          inputId = _this$state.inputId;
      return _react["default"].createElement(_InputGroup["default"], {
        focused: focused,
        invalid: invalid,
        quiet: quiet,
        disabled: disabled,
        className: (0, _classnames["default"])('spectrum-Stepper', 'react-spectrum-Stepper', {
          'spectrum-Stepper--quiet': quiet
        }, className),
        role: "group",
        "aria-label": ariaLabel || null,
        "aria-labelledby": ariaLabelledby || null
      }, _react["default"].createElement(_Textfield["default"], _extends({}, (0, _filterDOMProps["default"])(otherProps), {
        ref: function ref(t) {
          return _this2.textfield = t;
        },
        className: "spectrum-Stepper-input",
        id: inputId,
        value: value,
        defaultValue: defaultValue,
        type: "number",
        autoComplete: "off",
        "aria-label": ariaLabel || null,
        "aria-labelledby": ariaLabelledby || null,
        "aria-valuenow": value || null,
        "aria-valuetext": value || null,
        "aria-valuemin": min,
        "aria-valuemax": max,
        validationState: invalid || valueInvalid ? 'invalid' : null,
        step: step,
        placeholder: placeholder,
        disabled: disabled,
        readOnly: readOnly,
        quiet: quiet,
        onKeyDown: (0, _events.chain)(otherProps.onKeyDown, this.handleInputKeyDown),
        onFocus: (0, _events.chain)(otherProps.onFocus, this.handleFocus),
        onBlur: (0, _events.chain)(otherProps.onBlur, this.handleBlur),
        onChange: this.handleInputChange
      })), _react["default"].createElement("span", {
        className: "spectrum-Stepper-buttons",
        role: "presentation",
        onMouseDown: this.onMouseDown,
        onMouseUp: this.onMouseUp,
        onTouchStart: this.onTouchStart
      }, _react["default"].createElement(_Button["default"], {
        className: "spectrum-Stepper-stepUp",
        type: "button",
        tabIndex: "-1",
        "aria-controls": inputId,
        variant: "action",
        quiet: quiet,
        title: incrementTitle,
        "aria-label": incrementTitle,
        disabled: disabled || max != null && value >= max || readOnly,
        onClick: this.handleIncrementButtonClick,
        onMouseDown: function onMouseDown(e) {
          return e.preventDefault();
        },
        onMouseUp: function onMouseUp(e) {
          return e.preventDefault();
        },
        onFocus: this.handleFocus,
        onBlur: this.handleBlur
      }, _react["default"].createElement(_ChevronUpSmall["default"], {
        size: null,
        className: "spectrum-Stepper-stepUpIcon"
      })), _react["default"].createElement(_Button["default"], {
        className: "spectrum-Stepper-stepDown",
        type: "button",
        tabIndex: "-1",
        "aria-controls": inputId,
        variant: "action",
        quiet: quiet,
        title: decrementTitle,
        "aria-label": decrementTitle,
        disabled: disabled || min != null && value <= min || readOnly,
        onClick: this.handleDecrementButtonClick,
        onMouseDown: function onMouseDown(e) {
          return e.preventDefault();
        },
        onMouseUp: function onMouseUp(e) {
          return e.preventDefault();
        },
        onFocus: this.handleFocus,
        onBlur: this.handleBlur
      }, _react["default"].createElement(_ChevronDownSmall["default"], {
        size: null,
        className: "spectrum-Stepper-stepDownIcon"
      }))));
    }
  }]);

  return NumberInput;
}(_react.Component), _defineProperty(_class2, "propTypes", {
  /**
   * The default value.
   */
  defaultValue: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].number]),

  /**
   * Puts component into a controlled state.
   */
  value: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].number]),

  /**
   * The minimal value.
   */
  min: _propTypes["default"].number,

  /**
   * The maximum value.
   */
  max: _propTypes["default"].number,

  /**
   * The placeholder.
   */
  placeholder: _propTypes["default"].string,

  /**
   * Increment/decrement by step based on the current value. This differs from the w3 spec,
   * which will increment/decrement to the next multiple of the current step, regardless of current value.
   */
  step: _propTypes["default"].oneOfType([_propTypes["default"].number, _propTypes["default"].oneOf(['any'])]),

  /**
   * Whether the input is disabled.
   */
  disabled: _propTypes["default"].bool,

  /**
   * Whether the input is invalid. Will display a red border around the input.
   */
  invalid: _propTypes["default"].bool,

  /**
   * Whether the input is read only.
   */
  readOnly: _propTypes["default"].bool,

  /**
   * Should be a localized string, it's the tooltip and what a screenreader will announce for the decrement button.
   */
  decrementTitle: _propTypes["default"].string,

  /**
   * Should be a localized string, it's the tooltip and what a screenreader will announce for the increment button.
   */
  incrementTitle: _propTypes["default"].string,

  /**
   * The callback function when the input number is changed.
   */
  onChange: _propTypes["default"].func,

  /**
   * Whether the input should render using a quiet variant
   */
  quiet: _propTypes["default"].bool
}), _defineProperty(_class2, "defaultProps", {
  placeholder: formatMessage('Enter a number'),
  step: 1,
  disabled: false,
  invalid: false,
  readOnly: false,
  decrementTitle: formatMessage('Decrement'),
  incrementTitle: formatMessage('Increment'),
  onChange: function onChange() {}
}), _temp)) || _class) || _class;

exports["default"] = NumberInput;
