// @react/react-spectrum/Asset/js/Asset.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _AssetFile = _interopRequireDefault(require("./AssetFile"));

var _AssetFolder = _interopRequireDefault(require("./AssetFolder"));

var _AssetImage = _interopRequireDefault(require("./AssetImage"));

var _classnames = _interopRequireDefault(require("classnames"));

var _filterDOMProps = _interopRequireDefault(require("../../utils/filterDOMProps"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

if (process.env.SCALE_MEDIUM && process.env.SCALE_LARGE) {
  require("../../spectrum-css/asset/index.css");

  require("../../spectrum-css/asset/index-diff.css");
} else if (process.env.SCALE_LARGE) {
  require("../../spectrum-css/asset/index-lg.css");
} else {
  require("../../spectrum-css/asset/index.css");
}

if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_LIGHT) require("../../spectrum-css/asset/multiStops/light.css");
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_DARK) require("../../spectrum-css/asset/multiStops/dark.css");
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_LIGHTEST) require("../../spectrum-css/asset/multiStops/lightest.css");
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_DARKEST) require("../../spectrum-css/asset/multiStops/darkest.css");

var Asset =
/*#__PURE__*/
function (_React$Component) {
  _inherits(Asset, _React$Component);

  function Asset() {
    _classCallCheck(this, Asset);

    return _possibleConstructorReturn(this, _getPrototypeOf(Asset).apply(this, arguments));
  }

  _createClass(Asset, [{
    key: "render",
    value: function render(props) {
      var _this$props = this.props,
          type = _this$props.type,
          src = _this$props.src,
          placeholder = _this$props.placeholder,
          cache = _this$props.cache,
          headers = _this$props.headers,
          smartness = _this$props.smartness,
          alt = _this$props.alt,
          className = _this$props.className,
          onLoad = _this$props.onLoad,
          decorative = _this$props.decorative,
          otherProps = _objectWithoutProperties(_this$props, ["type", "src", "placeholder", "cache", "headers", "smartness", "alt", "className", "onLoad", "decorative"]);

      var content;

      if (type === 'image') {
        content = _react["default"].createElement(_AssetImage["default"], {
          src: src,
          placeholder: placeholder,
          smartness: smartness,
          onLoad: onLoad,
          cache: cache,
          headers: headers,
          alt: alt,
          decorative: decorative
        });
      } else if (type === 'folder') {
        content = _react["default"].createElement(_AssetFolder["default"], {
          alt: alt,
          decorative: decorative
        });
      } else {
        content = _react["default"].createElement(_AssetFile["default"], {
          alt: alt,
          decorative: decorative
        });
      }

      return _react["default"].createElement("div", _extends({}, (0, _filterDOMProps["default"])(otherProps), {
        className: (0, _classnames["default"])('spectrum-Asset', className)
      }), content);
    }
  }]);

  return Asset;
}(_react["default"].Component);

exports["default"] = Asset;

_defineProperty(Asset, "propTypes", {
  /** Asset can be an image, folder or file type */
  type: _propTypes["default"].oneOf(['image', 'folder', 'file']),

  /** Source of the Asset if it is an image */
  src: _propTypes["default"].string,

  /** Whether to use the image cache for the asset image */
  cache: _propTypes["default"].bool,

  /** HTTP headers to add to the request for the asset image */
  headers: _propTypes["default"].object,

  /** Placeholder image to display while the fullsize one is loading, if cached. */
  placeholder: _propTypes["default"].string,

  /** A factor between 0 and 1 that controls how the preview is scaled. When 0 ("dumb") the assets fits all the space available. */
  smartness: _propTypes["default"].number,

  /** Alternate content for screen readers */
  alt: _propTypes["default"].string,

  /** Whether the image is being used for decoration and should not be announced by screen readers */
  decorative: _propTypes["default"].bool,

  /** Load callback triggered when images load */
  onLoad: _propTypes["default"].func
});

_defineProperty(Asset, "defaultProps", {
  type: 'file',
  smartness: 1
});
