// @react/react-spectrum/Card/js/CardBody.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/**
 * Body of a card. Allows optional title, subtitle, and description props, and custom children displayed underneath.
 */
var CardBody =
/*#__PURE__*/
function (_React$Component) {
  _inherits(CardBody, _React$Component);

  function CardBody() {
    _classCallCheck(this, CardBody);

    return _possibleConstructorReturn(this, _getPrototypeOf(CardBody).apply(this, arguments));
  }

  _createClass(CardBody, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          title = _this$props.title,
          subtitle = _this$props.subtitle,
          description = _this$props.description,
          children = _this$props.children;
      var _this$context = this.context,
          cardVariant = _this$context.cardVariant,
          actionMenu = _this$context.actionMenu;
      var subtitleWrapper = subtitle ? _react["default"].createElement("div", {
        className: "spectrum-Card-subtitle"
      }, subtitle) : null;
      var descriptionWrapper = description ? _react["default"].createElement("div", {
        className: "spectrum-Card-description"
      }, description) : null;
      return _react["default"].createElement("div", {
        className: "spectrum-Card-body"
      }, title || actionMenu ? _react["default"].createElement("div", {
        className: "spectrum-Card-header"
      }, title ? _react["default"].createElement("div", {
        className: "spectrum-Card-title"
      }, title) : null, cardVariant === 'gallery' && subtitleWrapper, cardVariant === 'gallery' && descriptionWrapper, actionMenu ? _react["default"].createElement("div", {
        className: "spectrum-Card-actionButton"
      }, _react["default"].cloneElement(actionMenu, {
        alignRight: true
      })) : null) : null, cardVariant !== 'gallery' && (subtitle || description) ? _react["default"].createElement("div", {
        className: "spectrum-Card-content"
      }, subtitleWrapper, descriptionWrapper) : null, children);
    }
  }]);

  return CardBody;
}(_react["default"].Component);

exports["default"] = CardBody;

_defineProperty(CardBody, "displayName", 'CardBody');

_defineProperty(CardBody, "propTypes", {
  /** Title for the card */
  title: _propTypes["default"].string,

  /** Subtitle for the card, e.g. file type, displayed in capitalized style */
  subtitle: _propTypes["default"].string,

  /** Description for the card */
  description: _propTypes["default"].string
});

_defineProperty(CardBody, "contextTypes", {
  cardVariant: _propTypes["default"].string,
  actionMenu: _propTypes["default"].element
});
