// @react/react-spectrum/Card/js/Card.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _autobindDecorator = _interopRequireDefault(require("autobind-decorator"));

var _Checkbox = _interopRequireDefault(require("../../Checkbox/js/Checkbox"));

var _classnames = _interopRequireDefault(require("classnames"));

var _filterDOMProps = _interopRequireDefault(require("../../utils/filterDOMProps"));

var _intl = require("../../utils/intl");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

require("../style/index.css");

var _class, _class2, _temp;

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var intlMessages = {
  "cs-CZ": _interopRequireDefault(require("./../intl/cs-CZ.json"))["default"],
  "da-DK": _interopRequireDefault(require("./../intl/da-DK.json"))["default"],
  "de-DE": _interopRequireDefault(require("./../intl/de-DE.json"))["default"],
  "en-US": _interopRequireDefault(require("./../intl/en-US.json"))["default"],
  "es-ES": _interopRequireDefault(require("./../intl/es-ES.json"))["default"],
  "fi-FI": _interopRequireDefault(require("./../intl/fi-FI.json"))["default"],
  "fr-FR": _interopRequireDefault(require("./../intl/fr-FR.json"))["default"],
  "hu-HU": _interopRequireDefault(require("./../intl/hu-HU.json"))["default"],
  "it-IT": _interopRequireDefault(require("./../intl/it-IT.json"))["default"],
  "ja-JP": _interopRequireDefault(require("./../intl/ja-JP.json"))["default"],
  "ko-KR": _interopRequireDefault(require("./../intl/ko-KR.json"))["default"],
  "nb-NO": _interopRequireDefault(require("./../intl/nb-NO.json"))["default"],
  "nl-NL": _interopRequireDefault(require("./../intl/nl-NL.json"))["default"],
  "pl-PL": _interopRequireDefault(require("./../intl/pl-PL.json"))["default"],
  "pt-BR": _interopRequireDefault(require("./../intl/pt-BR.json"))["default"],
  "ru-RU": _interopRequireDefault(require("./../intl/ru-RU.json"))["default"],
  "sv-SE": _interopRequireDefault(require("./../intl/sv-SE.json"))["default"],
  "tr-TR": _interopRequireDefault(require("./../intl/tr-TR.json"))["default"],
  "uk-UA": _interopRequireDefault(require("./../intl/uk-UA.json"))["default"],
  "zh-CN": _interopRequireDefault(require("./../intl/zh-CN.json"))["default"],
  "zh-TW": _interopRequireDefault(require("./../intl/zh-TW.json"))["default"]
};

if (process.env.SCALE_MEDIUM && process.env.SCALE_LARGE) {
  require("../../spectrum-css/card/index.css");

  require("../../spectrum-css/card/index-diff.css");
} else if (process.env.SCALE_LARGE) {
  require("../../spectrum-css/card/index-lg.css");
} else {
  require("../../spectrum-css/card/index.css");
}

if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_LIGHT) require("../../spectrum-css/card/multiStops/light.css");
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_DARK) require("../../spectrum-css/card/multiStops/dark.css");
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_LIGHTEST) require("../../spectrum-css/card/multiStops/lightest.css");
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_DARKEST) require("../../spectrum-css/card/multiStops/darkest.css");

if (process.env.SCALE_MEDIUM && process.env.SCALE_LARGE) {
  require("../../spectrum-css/quickaction/index.css");

  require("../../spectrum-css/quickaction/index-diff.css");
} else if (process.env.SCALE_LARGE) {
  require("../../spectrum-css/quickaction/index-lg.css");
} else {
  require("../../spectrum-css/quickaction/index.css");
}

if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_LIGHT) require("../../spectrum-css/quickaction/multiStops/light.css");
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_DARK) require("../../spectrum-css/quickaction/multiStops/dark.css");
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_LIGHTEST) require("../../spectrum-css/quickaction/multiStops/lightest.css");
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_DARKEST) require("../../spectrum-css/quickaction/multiStops/darkest.css");
var formatMessage = (0, _intl.messageFormatter)(intlMessages);

var Card = (0, _autobindDecorator["default"])(_class = (_temp = _class2 =
/*#__PURE__*/
function (_React$Component) {
  _inherits(Card, _React$Component);

  function Card() {
    _classCallCheck(this, Card);

    return _possibleConstructorReturn(this, _getPrototypeOf(Card).apply(this, arguments));
  }

  _createClass(Card, [{
    key: "getChildContext",
    value: function getChildContext() {
      var actionMenu = this.props.actionMenu;

      if (this.props.size === 'S') {
        actionMenu = null;
      } // Find the CardBody to determine if it has a title.
      // This is used in CardPreview to decide whether to default the 
      // image to decorative for accessibility.


      var cardBody = _react["default"].Children.toArray(this.props.children).find(function (child) {
        return child.type.displayName === 'CardBody';
      });

      var hasTitle = Boolean(cardBody && cardBody.props.title);
      return {
        cardVariant: this.props.variant,
        cardSize: this.props.size,
        actionMenu: actionMenu,
        onLoad: this.props.onLoad,
        hasTitle: hasTitle
      };
    }
  }, {
    key: "handleCheckboxClick",
    value: function handleCheckboxClick(evt) {
      if (this.props.onSelectionChange) {
        evt.stopPropagation();
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          variant = _this$props.variant,
          size = _this$props.size,
          selected = _this$props.selected,
          allowsSelection = _this$props.allowsSelection,
          isDropTarget = _this$props.isDropTarget,
          quickActions = _this$props.quickActions,
          actionMenu = _this$props.actionMenu,
          children = _this$props.children,
          className = _this$props.className,
          otherProps = _objectWithoutProperties(_this$props, ["variant", "size", "selected", "allowsSelection", "isDropTarget", "quickActions", "actionMenu", "children", "className"]); // Small size is not supported for standard cards


      if (variant === 'standard' && size === 'S') {
        size = 'L';
      }

      var checkbox = null;

      if (allowsSelection) {
        checkbox = _react["default"].createElement("div", {
          className: (0, _classnames["default"])('spectrum-QuickActions', 'spectrum-Card-quickActions')
        }, _react["default"].createElement(_Checkbox["default"], {
          onChange: this.props.onSelectionChange,
          onClick: this.handleCheckboxClick,
          checked: selected,
          title: formatMessage('select')
        }));
      }

      if (quickActions && actionMenu) {
        throw new Error('Either quick actions or an action button can be passed to Card, not both.');
      }

      if (size === 'S' && actionMenu) {
        quickActions = _react["default"].createElement("div", {
          className: "spectrum-QuickActions spectrum-Card-actions"
        }, actionMenu);
      }

      var assetClass = (0, _classnames["default"])('spectrum-Card', "spectrum-Card--".concat(variant), {
        'spectrum-Card--large': size === 'L',
        'spectrum-Card--small': size === 'S',
        'is-selected': selected,
        'is-drop-target': isDropTarget
      }, className);
      return _react["default"].createElement("div", _extends({}, (0, _filterDOMProps["default"])(otherProps), {
        className: assetClass
      }), children, checkbox, quickActions);
    }
  }]);

  return Card;
}(_react["default"].Component), _defineProperty(_class2, "propTypes", {
  /** Card variant */
  variant: _propTypes["default"].oneOf(['standard', 'quiet', 'gallery']),

  /** Card can be large or small size */
  size: _propTypes["default"].oneOf(['S', 'L']),

  /** Whether or not the card supports selection */
  allowsSelection: _propTypes["default"].bool,

  /** Whether or not the card is selected */
  selected: _propTypes["default"].bool,

  /** What happens when the checkbox is clicked */
  onSelectionChange: _propTypes["default"].func,

  /** Whether the card is currently a drop target */
  isDropTarget: _propTypes["default"].bool,

  /** Quick actions to display on the card. */
  quickActions: _propTypes["default"].element,

  /** Action menu to display in the body of the card. */
  actionMenu: _propTypes["default"].element
}), _defineProperty(_class2, "defaultProps", {
  variant: 'standard',
  size: 'L',
  selected: false,
  allowsSelection: true
}), _defineProperty(_class2, "childContextTypes", {
  cardVariant: _propTypes["default"].string,
  cardSize: _propTypes["default"].string,
  actionMenu: _propTypes["default"].element,
  onLoad: _propTypes["default"].func,
  hasTitle: _propTypes["default"].bool
}), _temp)) || _class;

exports["default"] = Card;
