// @react/react-spectrum/Clock/js/Clock.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _autobindDecorator = _interopRequireDefault(require("autobind-decorator"));

var _number = require("../../utils/number");

var _classnames = _interopRequireDefault(require("classnames"));

var _convertUnsafeMethod = _interopRequireDefault(require("../../utils/convertUnsafeMethod"));

var _createId = _interopRequireDefault(require("../../utils/createId"));

var _filterDOMProps = _interopRequireDefault(require("../../utils/filterDOMProps"));

var _moment = require("../../utils/moment");

var _intl = require("../../utils/intl");

var _moment2 = _interopRequireDefault(require("moment"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireWildcard(require("react"));

var _Select = _interopRequireDefault(require("../../Select"));

var _Textfield = _interopRequireDefault(require("../../Textfield"));

require("../style/index.css");

var _VisuallyHidden = _interopRequireDefault(require("../../VisuallyHidden"));

var _class, _class2, _temp;

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; if (obj != null) { var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var intlMessages = {
  "cs-CZ": _interopRequireDefault(require("./../intl/cs-CZ.json"))["default"],
  "da-DK": _interopRequireDefault(require("./../intl/da-DK.json"))["default"],
  "de-DE": _interopRequireDefault(require("./../intl/de-DE.json"))["default"],
  "en-US": _interopRequireDefault(require("./../intl/en-US.json"))["default"],
  "es-ES": _interopRequireDefault(require("./../intl/es-ES.json"))["default"],
  "fi-FI": _interopRequireDefault(require("./../intl/fi-FI.json"))["default"],
  "fr-FR": _interopRequireDefault(require("./../intl/fr-FR.json"))["default"],
  "hu-HU": _interopRequireDefault(require("./../intl/hu-HU.json"))["default"],
  "it-IT": _interopRequireDefault(require("./../intl/it-IT.json"))["default"],
  "ja-JP": _interopRequireDefault(require("./../intl/ja-JP.json"))["default"],
  "ko-KR": _interopRequireDefault(require("./../intl/ko-KR.json"))["default"],
  "nb-NO": _interopRequireDefault(require("./../intl/nb-NO.json"))["default"],
  "nl-NL": _interopRequireDefault(require("./../intl/nl-NL.json"))["default"],
  "pl-PL": _interopRequireDefault(require("./../intl/pl-PL.json"))["default"],
  "pt-BR": _interopRequireDefault(require("./../intl/pt-BR.json"))["default"],
  "ru-RU": _interopRequireDefault(require("./../intl/ru-RU.json"))["default"],
  "sv-SE": _interopRequireDefault(require("./../intl/sv-SE.json"))["default"],
  "tr-TR": _interopRequireDefault(require("./../intl/tr-TR.json"))["default"],
  "uk-UA": _interopRequireDefault(require("./../intl/uk-UA.json"))["default"],
  "zh-CN": _interopRequireDefault(require("./../intl/zh-CN.json"))["default"],
  "zh-TW": _interopRequireDefault(require("./../intl/zh-TW.json"))["default"]
};
var formatMessage = (0, _intl.messageFormatter)(intlMessages);
/* In Firefox, input[type=number] always strips leading 0. */

var useTextInputType = typeof document !== 'undefined' ? 'MozAppearance' in document.documentElement.style : null;

var Clock = (0, _convertUnsafeMethod["default"])(_class = (0, _autobindDecorator["default"])(_class = (_temp = _class2 =
/*#__PURE__*/
function (_Component) {
  _inherits(Clock, _Component);

  function Clock(props) {
    var _this;

    _classCallCheck(this, Clock);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(Clock).call(this, props));
    _this._useTextInputType = useTextInputType;
    _this.clockId = (0, _createId["default"])();
    var _this$props = _this.props,
        value = _this$props.value,
        defaultValue = _this$props.defaultValue,
        displayFormat = _this$props.displayFormat,
        valueFormat = _this$props.valueFormat;

    _moment2["default"].locale((0, _intl.getLocale)());

    var val = (0, _moment.toMoment)(value || defaultValue || '', valueFormat);
    var isValid = val && val.isValid();
    var displayMeridiem = /a/i.test(displayFormat);
    _this.state = {
      value: val,
      hourText: isValid ? _this.getDisplayHour(val.hour(), displayMeridiem) : '',
      minuteText: isValid ? val.format('mm') : '',
      meridiemVal: _this.getMeridiemVal(val),
      displayMeridiem: displayMeridiem
    };
    return _this;
  }

  _createClass(Clock, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      var _this2 = this;

      if (this.props.autoFocus) {
        this.autoFocusTimeout = requestAnimationFrame(function () {
          return _this2.focus();
        });
      }
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      if (this.autoFocusTimeout) {
        cancelAnimationFrame(this.autoFocusTimeout);
      }
    }
  }, {
    key: "UNSAFE_componentWillReceiveProps",
    value: function UNSAFE_componentWillReceiveProps(nextProps) {
      if ('value' in nextProps) {
        this.setValue(nextProps.value, nextProps.valueFormat || this.props.valueFormat);
      }
    }
  }, {
    key: "setValue",
    value: function setValue(value, valueFormat) {
      var callback = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : function () {};
      var val = (0, _moment.toMoment)(value, valueFormat || this.props.valueFormat);
      var isValid = val && val.isValid();
      var _this$state = this.state,
          hourText = _this$state.hourText,
          minuteText = _this$state.minuteText,
          displayMeridiem = _this$state.displayMeridiem,
          meridiemVal = _this$state.meridiemVal;
      var newState = {
        value: val
      };

      if (!isValid) {
        newState.hourText = hourText || '';
        newState.minuteText = minuteText || '';
        newState.meridiemVal = meridiemVal || '';
      } else {
        var hourTextVal = this.getDisplayHour(val.hour(), displayMeridiem);
        var newMeridiemVal = this.getMeridiemVal(val);

        if (!hourText || +hourText !== hourTextVal) {
          newState.hourText = hourTextVal.padStart(2, '0');
        }

        if (!minuteText || +minuteText !== val.minute()) {
          newState.minuteText = val.format('mm').padStart(2, '0');
        }

        if (!meridiemVal || meridiemVal !== newMeridiemVal) {
          newState.meridiemVal = newMeridiemVal;
        }
      }

      this.setState(newState, callback);
    }
    /**
     * Handles change event on hour text input
     * @param {Event} e Change event
     * @private
     */

  }, {
    key: "handleHourChange",
    value: function handleHourChange(value, e) {
      var _this$state2 = this.state,
          minuteText = _this$state2.minuteText,
          meridiemVal = _this$state2.meridiemVal;
      e.stopPropagation();
      var newValue = value;

      if (value.replace) {
        newValue = value.replace(/\D/g, '');
      }

      this.changeTime(newValue, minuteText, meridiemVal);
    }
    /**
     * Handles change event on minute text input
     * @param {Event} e Change event
     * @private
     */

  }, {
    key: "handleMinuteChange",
    value: function handleMinuteChange(value, e) {
      var _this$state3 = this.state,
          hourText = _this$state3.hourText,
          meridiemVal = _this$state3.meridiemVal;
      e.stopPropagation();
      var newValue = value;

      if (value.replace) {
        newValue = value.replace(/\D/g, '');
      }

      this.changeTime(hourText, newValue, meridiemVal);
    }
    /**
     * Handles AM/PM Change
     * @param {string} value either 'am' or 'pm'
     * @private
     */

  }, {
    key: "handleMeridiemChange",
    value: function handleMeridiemChange(value) {
      var _this$state4 = this.state,
          hourText = _this$state4.hourText,
          minuteText = _this$state4.minuteText;
      this.changeTime(hourText, minuteText, value);
    }
    /**
     * Handles focus event on text input
     * @param {FocusEvent} e Focus event
     * @private
     */

  }, {
    key: "handleFocus",
    value: function handleFocus(e) {
      this.setState({
        focused: true
      });
    }
    /**
     * Handles blur event on hour text input
     * @param {FocusEvent} e Blur event
     * @private
     */

  }, {
    key: "handleHourBlur",
    value: function handleHourBlur(e) {
      var value = e.target.value;
      this.setState({
        hourText: value.padStart(2, '0'),
        focused: false
      });
    }
    /**
     * Handles blur event on minute text input
     * @param {FocusEvent} e Blur event
     * @private
     */

  }, {
    key: "handleMinuteBlur",
    value: function handleMinuteBlur(e) {
      var value = e.target.value;
      this.setState({
        minuteText: value.padStart(2, '0'),
        focused: false
      });
    }
    /**
     * Handles up or down arrow key event on text input, and depending
     * on value and direction loops time value and adjusts meridiem.
     * @param {Boolean} isHours Whether event target is the hours input
     * @param {KeyboardEvent} e Keyboard event
     * @private
     */

  }, {
    key: "handleTextfieldKeyDown",
    value: function handleTextfieldKeyDown() {
      var isHours = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : false;
      var e = arguments.length > 1 ? arguments[1] : undefined;
      var value = this.state.value;
      var newTime = value || (0, _moment.toMoment)('00:00', this.props.valueFormat);
      var _this$props2 = this.props,
          onChange = _this$props2.onChange,
          valueFormat = _this$props2.valueFormat;

      switch (e.key) {
        case 'ArrowUp':
        case 'Up':
          newTime = (0, _moment2["default"])(newTime).add(1, isHours ? 'hour' : 'minute');
          break;

        case 'ArrowDown':
        case 'Down':
          newTime = (0, _moment2["default"])(newTime).add(-1, isHours ? 'hour' : 'minute');
          break;

        default:
          return;
      }

      if (newTime !== value && _moment2["default"].isMoment(newTime)) {
        // don't keep the YYYY-MM-DD the same after adding/subtracting hours or minutes
        if (_moment2["default"].isMoment(value)) {
          newTime = newTime.year(value.year()).month(value.month()).date(value.date());
        }

        e.preventDefault();
        this.setValue(newTime, valueFormat, function () {
          onChange((0, _moment.formatMoment)(newTime, valueFormat), newTime.toDate());
        });
      }
    }
    /**
     * Updates time based on hour and minute text values.
     * @param {String} hourText   Hour text value
     * @param {String} minuteText Minute text value
     * @param {String} meridiemVal    'am' or 'pm'
     * @private
     */

  }, {
    key: "changeTime",
    value: function changeTime(hourText, minuteText, meridiemVal) {
      var _this$props3 = this.props,
          valueFormat = _this$props3.valueFormat,
          onChange = _this$props3.onChange;
      var _this$state5 = this.state,
          value = _this$state5.value,
          displayMeridiem = _this$state5.displayMeridiem; // Set the offset required to convert AM/PM times into 24 hr format

      var meridiemOffset = displayMeridiem && meridiemVal === 'pm' ? 12 : 0;
      var hours = parseInt(hourText, 10);

      if (hours < 12) {
        hours += meridiemOffset;
      } else if (displayMeridiem && meridiemVal === 'am' && hours === 12) {
        hours = 0;
      }

      var minutes = parseInt(minuteText, 10);
      var newTime = _moment2["default"].isMoment(value) && value.clone();

      if (isNaN(hours) || isNaN(minutes)) {
        newTime = '';
      } else {
        if (!_moment2["default"].isMoment(newTime)) {
          newTime = (0, _moment2["default"])();
        }

        newTime.hour((0, _number.clamp)(hours, 0, 23));
        newTime.minute((0, _number.clamp)(minutes, 0, 59));
        newTime.second(0);
        newTime.millisecond(0);
      }

      this.setState({
        hourText: hourText,
        minuteText: minuteText,
        meridiemVal: meridiemVal,
        newTime: newTime
      });

      if (!('value' in this.props)) {
        this.setState({
          value: newTime
        });
      }

      var validMoment = _moment2["default"].isMoment(newTime);

      if (newTime !== value) {
        onChange(validMoment ? (0, _moment.formatMoment)(newTime, valueFormat) : newTime, validMoment && newTime.toDate());
      }
    }
    /**
     * Sets focus to appropriate descendant.
     * @private
     */

  }, {
    key: "focus",
    value: function focus() {
      if (document.activeElement !== this.minuteRef) {
        this.hourRef && this.hourRef.focus();
      }
    }
    /**
     * Returns concatentated string of the ids for elements that label the clock field inputs.
     * @param {String} labelId Field label id
     * @param {String} groupId Fieldset id
     * @return {String}   aria-labelledby string
     * @private
     */

  }, {
    key: "getAriaLabelledbyForTextfield",
    value: function getAriaLabelledbyForTextfield(labelId, groupId) {
      var ariaLabel = this.props['aria-label'];
      var ariaLabelledby = this.props['aria-labelledby'];
      var ariaLabelledbyId = groupId;

      if (ariaLabel) {
        ariaLabelledbyId = groupId;
      }

      if (ariaLabelledby) {
        ariaLabelledbyId = ariaLabelledby;
      }

      return [ariaLabelledbyId, labelId].join(' ');
    }
    /**
     * Returns display hour for a given 24-hour value
     * @param {Number} hour Hour in range 0-23
     * @param {Boolean} displayMeridiem Output should be converted to 12-hour clock
     * @return {String} in range 1-12 (12-hour) or 0-23 (24-hour)
     * @private
     */

  }, {
    key: "getDisplayHour",
    value: function getDisplayHour(hour, displayMeridiem) {
      var newHour = displayMeridiem ? (hour + 11) % 12 + 1 : hour;
      return newHour.toString().padStart(2, '0');
    }
    /**
     * Returns localized label for AM/PM dropdown
     * @param {String} meridiem 'am' or 'pm'
     * @param {String} displayFormat from props
     * @return {String} in range 1-12 (12-hour) or 0-23 (24-hour)
     * @private
     */

  }, {
    key: "getMeridiemLabel",
    value: function getMeridiemLabel(meridiem, displayFormat) {
      var localizedMeridiem = formatMessage(meridiem);
      var upperCase = /A/.test(displayFormat);
      return upperCase ? localizedMeridiem.toUpperCase() : localizedMeridiem;
    }
    /**
     * Returns AM/PM (meridiem) value for a given moment
     * @param {Object} moment
     * @return {String} 'am', 'pm', or '' (for invalid moment values)
     * @private
     */

  }, {
    key: "getMeridiemVal",
    value: function getMeridiemVal(moment) {
      if (!moment || !moment.isValid()) {
        return '';
      } else {
        return moment.hour() >= 12 ? 'pm' : 'am';
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this3 = this;

      var _this$props4 = this.props,
          quiet = _this$props4.quiet,
          disabled = _this$props4.disabled,
          invalid = _this$props4.invalid,
          readOnly = _this$props4.readOnly,
          required = _this$props4.required,
          className = _this$props4.className,
          _this$props4$id = _this$props4.id,
          id = _this$props4$id === void 0 ? this.clockId : _this$props4$id,
          displayFormat = _this$props4.displayFormat,
          otherProps = _objectWithoutProperties(_this$props4, ["quiet", "disabled", "invalid", "readOnly", "required", "className", "id", "displayFormat"]);

      var _this$state6 = this.state,
          hourText = _this$state6.hourText,
          minuteText = _this$state6.minuteText,
          meridiemVal = _this$state6.meridiemVal,
          newTime = _this$state6.newTime,
          value = _this$state6.value,
          focused = _this$state6.focused,
          displayMeridiem = _this$state6.displayMeridiem;
      var hourMax = displayMeridiem ? 12 : 23;
      var hourMin = displayMeridiem ? 1 : 0;
      var groupId = id + '-group';
      var timeLabelId = id + '-time-label';
      var ariaLabel = otherProps['aria-label'];
      var ariaLabelledby = otherProps['aria-labelledby'];
      var groupAriaLabel = null;
      var groupAriaLabelledby = null;
      var formattedMoment = (0, _moment.formatMoment)(newTime || value, displayFormat);

      if (ariaLabel) {
        groupAriaLabel = ariaLabel;
        groupAriaLabelledby = [groupId, timeLabelId].join(' ');
        delete otherProps['aria-label'];
      } else {
        groupAriaLabelledby = timeLabelId;
      }

      if (ariaLabelledby) {
        groupAriaLabelledby = [ariaLabelledby, timeLabelId].join(' ');
        delete otherProps['aria-labelledby'];
      }

      delete otherProps.autoFocus;
      delete otherProps.valueFormat;
      delete otherProps.value;
      delete otherProps.defaultValue;
      var useTextInputType = this._useTextInputType;
      var inputType = useTextInputType ? 'text' : 'number';
      var inputRole = useTextInputType ? 'spinbutton' : null; // cant use input text with pattern to handle our validation
      // IE 11 requires that no two options in an alternation be able to start with the
      // same character (for the regex in a pattern attribute). If multiple options can
      // start with the same character (or blank), all but the first will be ignored
      // for that case.
      //
      // cant use input number by itself because IE 11 won't prevent other characters

      return _react["default"].createElement("div", _extends({
        className: (0, _classnames["default"])('react-spectrum-Clock', className),
        "aria-disabled": disabled,
        "aria-invalid": invalid,
        role: "group",
        id: groupId,
        "aria-label": groupAriaLabel,
        "aria-labelledby": groupAriaLabelledby
      }, (0, _filterDOMProps["default"])(otherProps)), _react["default"].createElement(_VisuallyHidden["default"], {
        element: "time",
        id: timeLabelId,
        "aria-live": focused ? 'polite' : 'off',
        hidden: !focused
      }, formattedMoment), _react["default"].createElement(_Textfield["default"], {
        ref: function ref(el) {
          return _this3.hourRef = el;
        },
        className: "react-spectrum-Clock-hour",
        type: inputType,
        role: inputRole,
        inputMode: "numeric",
        pattern: displayMeridiem ? '1[0-2]|0?[1-9]' : '2[0-3]|[01]?[0-9]',
        value: hourText,
        placeholder: displayMeridiem ? 'hh' : 'HH',
        min: !useTextInputType ? hourMin : null,
        max: !useTextInputType ? hourMax : null,
        "aria-valuemin": hourMin,
        "aria-valuemax": hourMax,
        "aria-valuenow": hourText || null,
        "aria-valuetext": hourText ? hourText.padStart(2, '0') : null,
        invalid: invalid,
        disabled: disabled,
        readOnly: readOnly,
        required: required,
        quiet: quiet,
        id: id,
        "aria-label": formatMessage('Hours'),
        "aria-labelledby": this.getAriaLabelledbyForTextfield(id, groupId),
        onKeyDown: this.handleTextfieldKeyDown.bind(this, true),
        onFocus: this.handleFocus,
        onChange: this.handleHourChange,
        onBlur: this.handleHourBlur
      }), _react["default"].createElement("span", {
        className: "react-spectrum-Clock-divider"
      }, ":"), _react["default"].createElement(_Textfield["default"], {
        ref: function ref(el) {
          return _this3.minuteRef = el;
        },
        className: "react-spectrum-Clock-minute",
        type: inputType,
        role: inputRole,
        inputMode: "numeric",
        pattern: "[0-5]?[0-9]",
        value: minuteText,
        placeholder: "mm",
        min: !useTextInputType ? 0 : null,
        max: !useTextInputType ? 59 : null,
        "aria-valuemin": 0,
        "aria-valuemax": 59,
        "aria-valuenow": minuteText || null,
        "aria-valuetext": minuteText ? minuteText.padStart(2, '0') : null,
        invalid: invalid,
        disabled: disabled,
        readOnly: readOnly,
        required: required,
        quiet: quiet,
        id: id + '-minutes',
        "aria-label": formatMessage('Minutes'),
        "aria-labelledby": this.getAriaLabelledbyForTextfield(id + '-minutes', groupId),
        onKeyDown: this.handleTextfieldKeyDown.bind(this, false),
        onFocus: this.handleFocus,
        onChange: this.handleMinuteChange,
        onBlur: this.handleMinuteBlur
      }), displayMeridiem && _react["default"].createElement(_Select["default"], {
        className: "react-spectrum-Clock-meridiem",
        id: id + '-meridiem',
        placeholder: formatMessage('AM/PM'),
        "aria-label": formatMessage('AM/PM'),
        "aria-labelledby": this.getAriaLabelledbyForTextfield(id + '-meridiem', groupId),
        onChange: this.handleMeridiemChange,
        options: [{
          label: this.getMeridiemLabel('am', displayFormat),
          value: 'am'
        }, {
          label: this.getMeridiemLabel('pm', displayFormat),
          value: 'pm'
        }],
        value: meridiemVal,
        disabled: disabled,
        alignRight: true,
        flexible: true
      }));
    }
  }]);

  return Clock;
}(_react.Component), _defineProperty(_class2, "displayName", 'Clock'), _defineProperty(_class2, "propTypes", {
  /**
   * Puts component into a controlled state. Clock can accept a number of different value representations
   */
  value: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].object, _propTypes["default"].number]),

  /**
   * Value component is initally set to. Clock can accept a number of different value representations
   */
  defaultValue: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].object, _propTypes["default"].number]),

  /**
   * Moment formats
   */
  valueFormat: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].func]),

  /**
   * Moment formats
   */
  displayFormat: _propTypes["default"].string,

  /**
   * Removes borders around inputs
   */
  quiet: _propTypes["default"].bool,

  /**
   * Greys out the control and makes it so you can't interact with it
   */
  disabled: _propTypes["default"].bool,

  /**
   * Marks the input red as invalid input
   */
  invalid: _propTypes["default"].bool,

  /**
   * Makes it so the value can't be changed by the user
   */
  readOnly: _propTypes["default"].bool,

  /**
   * Mark a form input as required, also necessary for the form to be submitted
   */
  required: _propTypes["default"].bool,

  /**
   * Boolean indicating this component gets focus when mounted to DOM
   */
  autoFocus: _propTypes["default"].bool,

  /**
   * Called when the value changes, returns a Moment object
   */
  onChange: _propTypes["default"].func
}), _defineProperty(_class2, "defaultProps", {
  valueFormat: 'HH:mm',
  displayFormat: 'HH:mm',
  quiet: false,
  disabled: false,
  invalid: false,
  readOnly: false,
  required: false,
  onChange: function onChange() {}
}), _temp)) || _class) || _class;

exports["default"] = Clock;
