// @react/react-spectrum/Alert/js/Alert.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = Alert;

var _AlertMedium = _interopRequireDefault(require("../../Icon/core/AlertMedium"));

var _Button = _interopRequireDefault(require("../../Button"));

var _classnames = _interopRequireDefault(require("classnames"));

var _filterDOMProps = _interopRequireDefault(require("../../utils/filterDOMProps"));

var _HelpMedium = _interopRequireDefault(require("../../Icon/core/HelpMedium"));

var _InfoMedium = _interopRequireDefault(require("../../Icon/core/InfoMedium"));

var _intl = require("../../utils/intl");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _SuccessMedium = _interopRequireDefault(require("../../Icon/core/SuccessMedium"));

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var intlMessages = {
  "cs-CZ": _interopRequireDefault(require("./../intl/cs-CZ.json"))["default"],
  "da-DK": _interopRequireDefault(require("./../intl/da-DK.json"))["default"],
  "de-DE": _interopRequireDefault(require("./../intl/de-DE.json"))["default"],
  "en-US": _interopRequireDefault(require("./../intl/en-US.json"))["default"],
  "es-ES": _interopRequireDefault(require("./../intl/es-ES.json"))["default"],
  "fi-FI": _interopRequireDefault(require("./../intl/fi-FI.json"))["default"],
  "fr-FR": _interopRequireDefault(require("./../intl/fr-FR.json"))["default"],
  "hu-HU": _interopRequireDefault(require("./../intl/hu-HU.json"))["default"],
  "it-IT": _interopRequireDefault(require("./../intl/it-IT.json"))["default"],
  "ja-JP": _interopRequireDefault(require("./../intl/ja-JP.json"))["default"],
  "ko-KR": _interopRequireDefault(require("./../intl/ko-KR.json"))["default"],
  "nb-NO": _interopRequireDefault(require("./../intl/nb-NO.json"))["default"],
  "nl-NL": _interopRequireDefault(require("./../intl/nl-NL.json"))["default"],
  "pl-PL": _interopRequireDefault(require("./../intl/pl-PL.json"))["default"],
  "pt-BR": _interopRequireDefault(require("./../intl/pt-BR.json"))["default"],
  "ru-RU": _interopRequireDefault(require("./../intl/ru-RU.json"))["default"],
  "sv-SE": _interopRequireDefault(require("./../intl/sv-SE.json"))["default"],
  "tr-TR": _interopRequireDefault(require("./../intl/tr-TR.json"))["default"],
  "uk-UA": _interopRequireDefault(require("./../intl/uk-UA.json"))["default"],
  "zh-CN": _interopRequireDefault(require("./../intl/zh-CN.json"))["default"],
  "zh-TW": _interopRequireDefault(require("./../intl/zh-TW.json"))["default"]
};

if (process.env.SCALE_MEDIUM && process.env.SCALE_LARGE) {
  require("../../spectrum-css/alert/index.css");

  require("../../spectrum-css/alert/index-diff.css");
} else if (process.env.SCALE_LARGE) {
  require("../../spectrum-css/alert/index-lg.css");
} else {
  require("../../spectrum-css/alert/index.css");
}

if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_LIGHT) require("../../spectrum-css/alert/multiStops/light.css");
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_DARK) require("../../spectrum-css/alert/multiStops/dark.css");
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_LIGHTEST) require("../../spectrum-css/alert/multiStops/lightest.css");
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_DARKEST) require("../../spectrum-css/alert/multiStops/darkest.css");
var ICONS = {
  error: _AlertMedium["default"],
  warning: _AlertMedium["default"],
  info: _InfoMedium["default"],
  help: _HelpMedium["default"],
  success: _SuccessMedium["default"]
};
var formatMessage = (0, _intl.messageFormatter)(intlMessages);

function Alert(_ref) {
  var header = _ref.header,
      children = _ref.children,
      closeLabel = _ref.closeLabel,
      onClose = _ref.onClose,
      _ref$variant = _ref.variant,
      variant = _ref$variant === void 0 ? 'info' : _ref$variant,
      className = _ref.className,
      _ref$alt = _ref.alt,
      alt = _ref$alt === void 0 ? formatMessage(variant || 'info') : _ref$alt,
      otherProps = _objectWithoutProperties(_ref, ["header", "children", "closeLabel", "onClose", "variant", "className", "alt"]);

  var AlertIcon = ICONS[variant];
  return _react["default"].createElement("div", _extends({
    className: (0, _classnames["default"])('spectrum-Alert', "spectrum-Alert--".concat(variant), className),
    role: "alert"
  }, (0, _filterDOMProps["default"])(otherProps)), _react["default"].createElement(AlertIcon, {
    size: null,
    className: "spectrum-Alert-icon",
    alt: alt
  }), _react["default"].createElement("div", {
    className: "spectrum-Alert-header"
  }, header), _react["default"].createElement("div", {
    className: "spectrum-Alert-content"
  }, children), closeLabel && _react["default"].createElement("div", {
    className: "spectrum-Alert-footer"
  }, _react["default"].createElement(_Button["default"], {
    label: closeLabel,
    onClick: onClose,
    quiet: true,
    variant: "primary"
  })));
}

Alert.propTypes = {
  /**
   * A string for the Header of the Alert
   */
  header: _propTypes["default"].string,

  /**
   * Any arbitrary node to render into content area
   */
  children: _propTypes["default"].node,

  /**
   * Affects the color and icon used by the Alert
   */
  variant: _propTypes["default"].oneOf(['error', 'warning', 'info', 'help', 'success']),

  /**
   * Label of the cancel button
   */
  closeLabel: _propTypes["default"].string,

  /**
   * Callback when dialog closes
   */
  onClose: _propTypes["default"].func
};
Alert.defaultProps = {
  variant: 'info'
};
Alert.displayName = 'Alert';
