// @react/react-spectrum/Breadcrumbs/js/Breadcrumbs.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _ChevronRightSmall = _interopRequireDefault(require("../../Icon/core/ChevronRightSmall"));

var _classnames = _interopRequireDefault(require("classnames"));

var _icon = require("../../utils/icon");

var _focusRing = require("../../utils/focusRing");

var _intl = require("../../utils/intl");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

require("../style/index.css");

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var intlMessages = {
  "cs-CZ": _interopRequireDefault(require("./../intl/cs-CZ.json"))["default"],
  "da-DK": _interopRequireDefault(require("./../intl/da-DK.json"))["default"],
  "de-DE": _interopRequireDefault(require("./../intl/de-DE.json"))["default"],
  "en-US": _interopRequireDefault(require("./../intl/en-US.json"))["default"],
  "es-ES": _interopRequireDefault(require("./../intl/es-ES.json"))["default"],
  "fi-FI": _interopRequireDefault(require("./../intl/fi-FI.json"))["default"],
  "fr-FR": _interopRequireDefault(require("./../intl/fr-FR.json"))["default"],
  "hu-HU": _interopRequireDefault(require("./../intl/hu-HU.json"))["default"],
  "it-IT": _interopRequireDefault(require("./../intl/it-IT.json"))["default"],
  "ja-JP": _interopRequireDefault(require("./../intl/ja-JP.json"))["default"],
  "ko-KR": _interopRequireDefault(require("./../intl/ko-KR.json"))["default"],
  "nb-NO": _interopRequireDefault(require("./../intl/nb-NO.json"))["default"],
  "nl-NL": _interopRequireDefault(require("./../intl/nl-NL.json"))["default"],
  "pl-PL": _interopRequireDefault(require("./../intl/pl-PL.json"))["default"],
  "pt-BR": _interopRequireDefault(require("./../intl/pt-BR.json"))["default"],
  "ru-RU": _interopRequireDefault(require("./../intl/ru-RU.json"))["default"],
  "sv-SE": _interopRequireDefault(require("./../intl/sv-SE.json"))["default"],
  "tr-TR": _interopRequireDefault(require("./../intl/tr-TR.json"))["default"],
  "uk-UA": _interopRequireDefault(require("./../intl/uk-UA.json"))["default"],
  "zh-CN": _interopRequireDefault(require("./../intl/zh-CN.json"))["default"],
  "zh-TW": _interopRequireDefault(require("./../intl/zh-TW.json"))["default"]
};

if (process.env.SCALE_MEDIUM && process.env.SCALE_LARGE) {
  require("../../spectrum-css/breadcrumb/index.css");

  require("../../spectrum-css/breadcrumb/index-diff.css");
} else if (process.env.SCALE_LARGE) {
  require("../../spectrum-css/breadcrumb/index-lg.css");
} else {
  require("../../spectrum-css/breadcrumb/index.css");
}

if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_LIGHT) require("../../spectrum-css/breadcrumb/multiStops/light.css");
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_DARK) require("../../spectrum-css/breadcrumb/multiStops/dark.css");
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_LIGHTEST) require("../../spectrum-css/breadcrumb/multiStops/lightest.css");
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_DARKEST) require("../../spectrum-css/breadcrumb/multiStops/darkest.css");
var formatMessage = (0, _intl.messageFormatter)(intlMessages);

var Breadcrumbs =
/*#__PURE__*/
function (_React$Component) {
  _inherits(Breadcrumbs, _React$Component);

  function Breadcrumbs() {
    _classCallCheck(this, Breadcrumbs);

    return _possibleConstructorReturn(this, _getPrototypeOf(Breadcrumbs).apply(this, arguments));
  }

  _createClass(Breadcrumbs, [{
    key: "handleFocus",
    value: function handleFocus(e) {
      e.target.classList.add(_focusRing.FOCUS_RING_CLASSNAME);
    }
  }, {
    key: "handleBlur",
    value: function handleBlur(e) {
      e.target.classList.remove(_focusRing.FOCUS_RING_CLASSNAME);
    }
  }, {
    key: "renderLink",
    value: function renderLink(item, i) {
      var _this$props = this.props,
          items = _this$props.items,
          renderLink = _this$props.renderLink;
      var isHref = !!item.href;
      var isCurrent = this.isCurrent(i);
      var hasItems = items.length > 1;
      var linkProps = {
        className: 'spectrum-Breadcrumbs-itemLink',
        role: !isHref ? 'link' : null,
        href: !isCurrent ? item.href : null,
        target: !isCurrent && isHref ? '_self' : null,
        onClick: hasItems && !isCurrent ? this.props.onBreadcrumbClick.bind(null, item, items.length - i - 1) : undefined,
        onFocus: hasItems && !isCurrent ? this.handleFocus.bind(this) : null,
        onBlur: hasItems && !isCurrent ? this.handleBlur.bind(this) : null,
        'aria-current': isCurrent ? 'page' : null,
        tabIndex: !isHref && !isCurrent ? 0 : null
      };

      if (renderLink) {
        return renderLink(_objectSpread({}, linkProps, {
          label: item.label
        }));
      }

      var Element = !isHref ? 'div' : 'a';
      return _react["default"].createElement(Element, linkProps, item.label);
    }
  }, {
    key: "isCurrent",
    value: function isCurrent(i) {
      return i === this.props.items.length - 1;
    }
  }, {
    key: "render",
    value: function render() {
      var _this = this;

      var _this$props2 = this.props,
          items = _this$props2.items,
          icon = _this$props2.icon,
          className = _this$props2.className,
          variant = _this$props2.variant,
          ariaLevel = _this$props2.ariaLevel,
          id = _this$props2.id,
          otherProps = _objectWithoutProperties(_this$props2, ["items", "icon", "className", "variant", "ariaLevel", "id"]);

      var isTitleVariant = variant === Breadcrumbs.variant.TITLE;

      if (!otherProps['aria-label']) {
        otherProps['aria-label'] = formatMessage('Breadcrumbs');
      }

      return _react["default"].createElement("nav", {
        id: id,
        "aria-label": otherProps['aria-label'],
        "aria-labelledby": otherProps['aria-labelledby']
      }, (0, _icon.cloneIcon)(icon, {
        size: 'S',
        className: 'react-spectrum-Breadcrumbs-icon'
      }), _react["default"].createElement("ul", {
        className: (0, _classnames["default"])('spectrum-Breadcrumbs', {
          'spectrum-Breadcrumbs--title': isTitleVariant
        }, className)
      }, items.map(function (item, i) {
        return _react["default"].createElement("li", {
          key: "spectrum-Breadcrumb-".concat(i),
          className: "spectrum-Breadcrumbs-item"
        }, _this.isCurrent(i) && isTitleVariant ? _react["default"].createElement("h1", {
          className: "spectrum-Heading--pageTitle",
          "aria-level": ariaLevel
        }, _this.renderLink(item, i)) : _this.renderLink(item, i), !_this.isCurrent(i) && _react["default"].createElement(_ChevronRightSmall["default"], {
          size: null,
          className: "spectrum-Breadcrumbs-itemSeparator"
        }));
      })));
    }
  }]);

  return Breadcrumbs;
}(_react["default"].Component);

exports["default"] = Breadcrumbs;

_defineProperty(Breadcrumbs, "variant", {
  NONE: 'none',
  TITLE: 'title'
});

_defineProperty(Breadcrumbs, "propTypes", {
  /**
   * Array of item objects included in list of breadcrumbs.
   * Each item must have a label string property, which will serve as the rendered text for the breadcrumb.
   * Each item may also have an href string property, which should be a valid URL to open the breadcrumb location.
   */
  items: _propTypes["default"].arrayOf(_propTypes["default"].oneOfType([_propTypes["default"].shape({
    href: _propTypes["default"].string,
    label: _propTypes["default"].string.isRequired
  }), _propTypes["default"].shape({
    label: _propTypes["default"].string.isRequired
  })])),

  /**
   * For best results, use a React Spectrum Icon
   */
  icon: _propTypes["default"].node,

  /**
   * Will set the last breadcrumb on a new line and give it a <h> tag
   */
  variant: _propTypes["default"].oneOf([Breadcrumbs.variant.NONE, Breadcrumbs.variant.TITLE]),

  /**
   * Will not change the appearance
   */
  ariaLevel: _propTypes["default"].number,

  /**
   * Called when a breadcrumb is clicked with an object containing the label of the clicked breadcrumb
   * @callback Breadcrumbs~onBreadcrumbClick
   * @param {Object} event - Event object
   * @param {string} event.label - label of breadcrumb clicked
   */
  onBreadcrumbClick: _propTypes["default"].func,

  /**
   * A function that returns a href wrapper component.
   * Useful in providing custom href component(eg. Link from react-router-dom).
   *
   * ```js
   *  <Breadcrumbs renderLink={(props) => <Link {...props} to="/">Foo</Link>} items={items} />
   * ```
   */
  renderLink: _propTypes["default"].func
});

_defineProperty(Breadcrumbs, "defaultProps", {
  items: [],
  icon: null,
  variant: Breadcrumbs.variant.NONE,
  ariaLevel: null,
  onBreadcrumbClick: function onBreadcrumbClick() {}
});
