// @react/react-spectrum/TagList/js/Tag.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _autobindDecorator = _interopRequireDefault(require("autobind-decorator"));

var _Avatar = _interopRequireDefault(require("../../Avatar"));

var _Button = _interopRequireDefault(require("../../Button"));

var _classnames = _interopRequireDefault(require("classnames"));

var _icon = require("../../utils/icon");

var _CrossSmall = _interopRequireDefault(require("../../Icon/core/CrossSmall"));

var _filterDOMProps = _interopRequireDefault(require("../../utils/filterDOMProps"));

var _focusRing = require("../../utils/focusRing");

var _intl = require("../../utils/intl");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _class, _class2, _temp;

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var intlMessages = {
  "cs-CZ": _interopRequireDefault(require("./../intl/cs-CZ.json"))["default"],
  "da-DK": _interopRequireDefault(require("./../intl/da-DK.json"))["default"],
  "de-DE": _interopRequireDefault(require("./../intl/de-DE.json"))["default"],
  "en-US": _interopRequireDefault(require("./../intl/en-US.json"))["default"],
  "es-ES": _interopRequireDefault(require("./../intl/es-ES.json"))["default"],
  "fi-FI": _interopRequireDefault(require("./../intl/fi-FI.json"))["default"],
  "fr-FR": _interopRequireDefault(require("./../intl/fr-FR.json"))["default"],
  "hu-HU": _interopRequireDefault(require("./../intl/hu-HU.json"))["default"],
  "it-IT": _interopRequireDefault(require("./../intl/it-IT.json"))["default"],
  "ja-JP": _interopRequireDefault(require("./../intl/ja-JP.json"))["default"],
  "ko-KR": _interopRequireDefault(require("./../intl/ko-KR.json"))["default"],
  "nb-NO": _interopRequireDefault(require("./../intl/nb-NO.json"))["default"],
  "nl-NL": _interopRequireDefault(require("./../intl/nl-NL.json"))["default"],
  "pl-PL": _interopRequireDefault(require("./../intl/pl-PL.json"))["default"],
  "pt-BR": _interopRequireDefault(require("./../intl/pt-BR.json"))["default"],
  "ru-RU": _interopRequireDefault(require("./../intl/ru-RU.json"))["default"],
  "sv-SE": _interopRequireDefault(require("./../intl/sv-SE.json"))["default"],
  "tr-TR": _interopRequireDefault(require("./../intl/tr-TR.json"))["default"],
  "uk-UA": _interopRequireDefault(require("./../intl/uk-UA.json"))["default"],
  "zh-CN": _interopRequireDefault(require("./../intl/zh-CN.json"))["default"],
  "zh-TW": _interopRequireDefault(require("./../intl/zh-TW.json"))["default"]
};

if (process.env.SCALE_MEDIUM && process.env.SCALE_LARGE) {
  require("../../spectrum-css/tags/index.css");

  require("../../spectrum-css/tags/index-diff.css");
} else if (process.env.SCALE_LARGE) {
  require("../../spectrum-css/tags/index-lg.css");
} else {
  require("../../spectrum-css/tags/index.css");
}

if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_LIGHT) require("../../spectrum-css/tags/multiStops/light.css");
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_DARK) require("../../spectrum-css/tags/multiStops/dark.css");
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_LIGHTEST) require("../../spectrum-css/tags/multiStops/lightest.css");
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_DARKEST) require("../../spectrum-css/tags/multiStops/darkest.css");
var formatMessage = (0, _intl.messageFormatter)(intlMessages);
/**
 * A tag is used to categorize content and display filters.
 */

var Tag = (0, _autobindDecorator["default"])(_class = (_temp = _class2 =
/*#__PURE__*/
function (_React$Component) {
  _inherits(Tag, _React$Component);

  function Tag(props) {
    var _this;

    _classCallCheck(this, Tag);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(Tag).call(this, props));
    _this.state = {
      tagFocused: false
    };
    return _this;
  }

  _createClass(Tag, [{
    key: "handleButtonFocus",
    value: function handleButtonFocus(e) {
      this.setState({
        tagFocused: true
      });
    }
  }, {
    key: "handleButtonBlur",
    value: function handleButtonBlur(e) {
      this.setState({
        tagFocused: false
      });
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          value = _this$props.value,
          children = _this$props.children,
          avatar = _this$props.avatar,
          icon = _this$props.icon,
          _this$props$closable = _this$props.closable,
          closable = _this$props$closable === void 0 ? false : _this$props$closable,
          _this$props$disabled = _this$props.disabled,
          disabled = _this$props$disabled === void 0 ? false : _this$props$disabled,
          _this$props$selected = _this$props.selected,
          selected = _this$props$selected === void 0 ? false : _this$props$selected,
          _this$props$invalid = _this$props.invalid,
          invalid = _this$props$invalid === void 0 ? false : _this$props$invalid,
          className = _this$props.className,
          _this$props$onClose = _this$props.onClose,
          onClose = _this$props$onClose === void 0 ? function () {} : _this$props$onClose,
          tabIndex = _this$props.tabIndex,
          role = _this$props.role,
          otherProps = _objectWithoutProperties(_this$props, ["value", "children", "avatar", "icon", "closable", "disabled", "selected", "invalid", "className", "onClose", "tabIndex", "role"]);

      var tag = this.tag;
      var tagFocused = this.state.tagFocused;
      var removeString = formatMessage('Remove');
      var childContent = children || value;
      var ariaLabel = childContent ? "".concat(removeString, ": ").concat(childContent) : {
        removeString: removeString
      };

      function handleKeyDown(e) {
        switch (e.keyCode) {
          case 46: // delete

          case 8: // backspace

          case 32:
            // space
            onClose(value || children, e);
            e.preventDefault();
            break;
        }
      }

      function handleButtonClick(e) {
        onClose(value || children, e); // If the button is clicked and this is a gridcell it must be a mouse event
        // Set focus to the tag rather than the button as that is where focus manager
        // expects it to be

        if (role === 'gridcell') {
          if (tag) {
            tag.focus();
          }
        }
      }

      return _react["default"].createElement("div", _extends({
        ref: function ref(t) {
          _this2.tag = t;
        },
        className: (0, _classnames["default"])('spectrum-Tags-item', {
          'is-selected': selected,
          'is-disabled': disabled,
          'spectrum-Tags-item--deletable': closable,
          'is-invalid': invalid
        }, _defineProperty({}, _focusRing.FOCUS_RING_CLASSNAME, tagFocused), className),
        tabIndex: tabIndex,
        "aria-selected": !disabled && selected,
        onKeyDown: !disabled && closable ? function (e) {
          handleKeyDown(e);
        } : null
      }, (0, _filterDOMProps["default"])(otherProps), {
        role: role === 'gridcell' ? 'row' : undefined
      }), avatar && _react["default"].createElement(_Avatar["default"], {
        alt: "",
        src: avatar,
        "aria-hidden": "true"
      }), (0, _icon.cloneIcon)(icon, {
        size: 'S',
        className: 'spectrum-Tags-itemIcon'
      }), _react["default"].createElement("span", {
        role: role,
        className: "spectrum-Tags-itemLabel"
      }, childContent), closable && _react["default"].createElement("span", {
        role: role
      }, _react["default"].createElement(_Button["default"], {
        type: "button",
        tabIndex: role === 'gridcell' || disabled ? '-1' : undefined,
        "aria-label": ariaLabel,
        className: "spectrum-ClearButton--small",
        variant: "clear",
        icon: _react["default"].createElement(_CrossSmall["default"], null),
        title: removeString,
        onClick: !disabled ? function (e) {
          handleButtonClick(e);
        } : null,
        onBlur: this.handleButtonBlur,
        onFocus: this.handleButtonFocus
      })));
    }
  }]);

  return Tag;
}(_react["default"].Component), _defineProperty(_class2, "propTypes", {
  /** Avatar to use in the tag */
  avatar: _propTypes["default"].string,

  /** Whether the tag is removable from the tag list */
  closeable: _propTypes["default"].bool,

  /** Function to be executed upon tag closure */
  onClose: _propTypes["default"].func,

  /** Whether the tag is disabled from user interaction */
  disabled: _propTypes["default"].bool,

  /** Icon to use in the tag */
  icon: _propTypes["default"].object,

  /** Whether the tag is selected */
  selected: _propTypes["default"].bool,

  /** Whether the tag ought to be colored red to reflect invalid status */
  invalid: _propTypes["default"].bool
}), _defineProperty(_class2, "defaultProps", {
  closeable: false,
  disabled: false,
  selected: false
}), _temp)) || _class;

exports["default"] = Tag;
