// @react/react-spectrum/QuickActions/js/QuickActionsItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.QuickActionsItem = QuickActionsItem;

var _Button = _interopRequireDefault(require("../../Button"));

var _classnames = _interopRequireDefault(require("classnames"));

var _Menu = require("../../Menu");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

/**
 * A quick actions item
 */
function QuickActionsItem(_ref) {
  var className = _ref.className,
      icon = _ref.icon,
      label = _ref.label,
      onClick = _ref.onClick,
      isMenuItem = _ref.isMenuItem,
      ariaLabel = _ref['aria-label'],
      otherProps = _objectWithoutProperties(_ref, ["className", "icon", "label", "onClick", "isMenuItem", "aria-label"]);

  return isMenuItem ? _react["default"].createElement(_Menu.MenuItem, _extends({}, otherProps, {
    className: "spectrum-".concat(QuickActionsItem.displayName),
    icon: icon,
    onClick: onClick,
    "aria-label": ariaLabel
  }), label) : _react["default"].createElement(_Button["default"], _extends({}, otherProps, {
    className: (0, _classnames["default"])(className, "spectrum-".concat(QuickActionsItem.displayName)),
    variant: "action",
    quiet: true,
    icon: icon,
    label: icon ? undefined : label,
    "aria-label": icon ? ariaLabel || label : ariaLabel,
    onClick: onClick,
    role: "menuitem"
  }));
}

QuickActionsItem.displayName = 'QuickActionsItem';
QuickActionsItem.propTypes = {
  /**
   * Custom className to apply to this component.
   */
  className: _propTypes["default"].string,

  /**
   * An icon to render in the quick actions item.
   */
  icon: _propTypes["default"].element,

  /**
   * The label to display in the quick actions item.
   */
  label: _propTypes["default"].string.isRequired,

  /**
   * Callback for when the quick action item is clicked.
   */
  onClick: _propTypes["default"].func
};
