// @react/react-spectrum/Shell/js/ShellUserProfile.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = ShellUserProfile;

var _Button = _interopRequireDefault(require("../../Button"));

var _classnames = _interopRequireDefault(require("classnames"));

var _react = _interopRequireDefault(require("react"));

var _ShellMenu = _interopRequireDefault(require("./ShellMenu"));

require("../style/ShellUserProfile.css");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function ShellUserProfile(_ref) {
  var name = _ref.name,
      heading = _ref.heading,
      subheading = _ref.subheading,
      _ref$avatarUrl = _ref.avatarUrl,
      avatarUrl = _ref$avatarUrl === void 0 ? '#' : _ref$avatarUrl,
      _ref$profileUrl = _ref.profileUrl,
      profileUrl = _ref$profileUrl === void 0 ? '#' : _ref$profileUrl,
      _ref$onSignOut = _ref.onSignOut,
      onSignOut = _ref$onSignOut === void 0 ? function () {} : _ref$onSignOut,
      children = _ref.children,
      className = _ref.className,
      otherProps = _objectWithoutProperties(_ref, ["name", "heading", "subheading", "avatarUrl", "profileUrl", "onSignOut", "children", "className"]);

  return _react["default"].createElement(_ShellMenu["default"], _extends({
    placement: "right",
    animateFrom: "top",
    target: _react["default"].createElement(_Button["default"], {
      className: "coral3-Shell-menu-button",
      variant: "minimal"
    }, _react["default"].createElement("img", {
      src: avatarUrl,
      alt: "",
      "aria-hidden": "true"
    }))
  }, otherProps), _react["default"].createElement("div", {
    className: (0, _classnames["default"])('coral3-Shell-user', className)
  }, _react["default"].createElement("div", {
    className: "coral3-Shell-user-container"
  }, _react["default"].createElement("div", {
    className: "coral3-Shell-user-image"
  }, _react["default"].createElement("img", {
    src: avatarUrl,
    alt: "",
    "aria-hidden": "true",
    className: "coral3-Shell-user-avatar"
  })), _react["default"].createElement("div", {
    className: "coral3-Shell-user-name"
  }, name), _react["default"].createElement("div", {
    className: "coral3-Shell-user-heading"
  }, heading && heading !== 'null' ? heading : ''), _react["default"].createElement("div", {
    className: "coral3-Shell-user-subheading"
  }, subheading && subheading !== 'null' ? subheading : '')), _react["default"].createElement("div", {
    className: "coral3-Shell-user-content"
  }, children), _react["default"].createElement("div", {
    className: "coral3-Shell-user-footer"
  }, _react["default"].createElement(_Button["default"], {
    element: "a",
    variant: "minimal",
    href: profileUrl
  }, "Edit Profile"), _react["default"].createElement(_Button["default"], {
    element: "a",
    variant: "minimal",
    onClick: onSignOut
  }, "Sign Out"))));
}

ShellUserProfile.displayName = 'ShellUserProfile';
