// @react/react-spectrum/Shell/js/ShellOrgSwitcher.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _Button = _interopRequireDefault(require("../../Button"));

var _classnames = _interopRequireDefault(require("classnames"));

var _convertUnsafeMethod = _interopRequireDefault(require("../../utils/convertUnsafeMethod"));

var _react = _interopRequireWildcard(require("react"));

var _Search = _interopRequireDefault(require("../../Search"));

var _SelectList = _interopRequireDefault(require("../../SelectList"));

var _ShellMenu = _interopRequireDefault(require("./ShellMenu"));

require("../style/ShellOrgSwitcher.css");

var _class, _class2, _temp;

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; if (obj != null) { var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var ShellOrgSwitcher = (0, _convertUnsafeMethod["default"])(_class = (_temp = _class2 =
/*#__PURE__*/
function (_Component) {
  _inherits(ShellOrgSwitcher, _Component);

  function ShellOrgSwitcher() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, ShellOrgSwitcher);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(ShellOrgSwitcher)).call.apply(_getPrototypeOf2, [this].concat(args)));

    _defineProperty(_assertThisInitialized(_this), "state", {
      searchTerm: '',
      visibleOptions: _this.filterVisibleOptions(_this.props.options, '')
    });

    _defineProperty(_assertThisInitialized(_this), "handleVisible", function () {
      if (_this.contentRef) {
        _this.contentRef.querySelector('.spectrum-Search-input').focus();
      }
    });

    _defineProperty(_assertThisInitialized(_this), "handleSearchChange", function (searchTerm) {
      var options = _this.props.options;

      _this.setState({
        searchTerm: searchTerm,
        visibleOptions: _this.filterVisibleOptions(options, searchTerm)
      });
    });

    return _this;
  }

  _createClass(ShellOrgSwitcher, [{
    key: "UNSAFE_componentWillReceiveProps",
    value: function UNSAFE_componentWillReceiveProps(nextProps) {
      if (nextProps.options !== this.props.options) {
        this.setState({
          visibleOptions: this.filterVisibleOptions(nextProps.options, this.state.searchTerm)
        });
      }
    }
  }, {
    key: "filterVisibleOptions",
    value: function filterVisibleOptions(options, searchTerm) {
      return options.filter(function (option) {
        return option.label.toLowerCase().indexOf(searchTerm) !== -1;
      });
    }
  }, {
    key: "getSelectedLabel",
    value: function getSelectedLabel() {
      var _this$props = this.props,
          options = _this$props.options,
          value = _this$props.value;
      var selectedOptions = options.filter(function (option) {
        return option.value === value;
      });

      if (selectedOptions.length) {
        return selectedOptions[0].label;
      }

      return value || '';
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props2 = this.props,
          value = _this$props2.value,
          className = _this$props2.className,
          manageOrgsUrl = _this$props2.manageOrgsUrl,
          onOrgChange = _this$props2.onOrgChange,
          otherProps = _objectWithoutProperties(_this$props2, ["value", "className", "manageOrgsUrl", "onOrgChange"]);

      var visibleOptions = this.state.visibleOptions;
      return _react["default"].createElement(_ShellMenu["default"], _extends({
        placement: "right",
        animateFrom: "top",
        target: _react["default"].createElement(_Button["default"], {
          variant: "minimal",
          className: "coral3-Shell-menu-button"
        }, this.getSelectedLabel()),
        onVisible: this.handleVisible
      }, otherProps), _react["default"].createElement("div", {
        ref: function ref(el) {
          _this2.contentRef = el;
        },
        className: (0, _classnames["default"])('coral3-Shell-orgSwitcher', className)
      }, _react["default"].createElement(_Search["default"], {
        className: "coral3-Shell-orgSwitcher-search",
        placeholder: "Search Organizations",
        onChange: this.handleSearchChange,
        quiet: true
      }), visibleOptions.length !== 0 && _react["default"].createElement(_SelectList["default"], {
        className: "coral3-Shell-orgSwitcher-items",
        value: value,
        options: visibleOptions,
        onChange: onOrgChange
      }), visibleOptions.length === 0 && _react["default"].createElement("div", {
        className: "coral3-Shell-orgSwitcher-resultMessage"
      }, _react["default"].createElement("div", {
        className: "coral3-Shell-orgSwitcher-resultMessage-container"
      }, _react["default"].createElement("div", {
        className: "coral-Heading--1 coral3-Shell-orgSwitcher-resultMessage-heading"
      }, "No organizations found."))), _react["default"].createElement("div", {
        className: "coral3-Shell-orgSwitcher-footer"
      }, _react["default"].createElement(_Button["default"], {
        element: "a",
        variant: "minimal",
        href: manageOrgsUrl,
        block: true
      }, "Manage Organizations"))));
    }
  }]);

  return ShellOrgSwitcher;
}(_react.Component), _defineProperty(_class2, "defaultProps", {
  value: '',
  manageOrgsUrl: '#',
  options: [],
  onOrgChange: function onOrgChange() {}
}), _temp)) || _class;

exports["default"] = ShellOrgSwitcher;
ShellOrgSwitcher.displayName = 'ShellOrgSwitcher';
