// @react/react-spectrum/DropdownButton/js/DropdownButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = DropdownButton;

var _Button = _interopRequireDefault(require("../../Button"));

var _Dropdown = _interopRequireDefault(require("../../Dropdown"));

var _Menu = require("../../Menu");

var _More = _interopRequireDefault(require("../../Icon/More"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function DropdownButton(_ref) {
  var alignRight = _ref.alignRight,
      onClose = _ref.onClose,
      onOpen = _ref.onOpen,
      onSelect = _ref.onSelect,
      children = _ref.children,
      closeOnSelect = _ref.closeOnSelect,
      menuClassName = _ref.menuClassName,
      holdAffordance = _ref.holdAffordance,
      onClick = _ref.onClick,
      props = _objectWithoutProperties(_ref, ["alignRight", "onClose", "onOpen", "onSelect", "children", "closeOnSelect", "menuClassName", "holdAffordance", "onClick"]);

  var trigger = holdAffordance ? 'longClick' : undefined;
  return _react["default"].createElement(_Dropdown["default"], {
    style: {
      display: 'inline-block'
    },
    onClose: onClose,
    onOpen: onOpen,
    onClick: onClick,
    onSelect: onSelect,
    alignRight: alignRight,
    closeOnSelect: closeOnSelect,
    trigger: trigger
  }, _react["default"].createElement(_Button["default"], _extends({
    variant: "action",
    quiet: true,
    holdAffordance: holdAffordance,
    icon: _react["default"].createElement(_More["default"], {
      alt: "\u2026"
    })
  }, props)), _react["default"].createElement(_Menu.Menu, {
    className: menuClassName,
    dropdownMenu: true
  }, children));
}

DropdownButton.propTypes = {
  /**
   * If true, dropdown will close on selection of an item
   */
  closeOnSelect: _propTypes["default"].bool,

  /**
   * Callback for when the dropdown is opened
   */
  onOpen: _propTypes["default"].func,

  /**
   * Callback for when the dropdown is closed
   */
  onClose: _propTypes["default"].func,

  /**
   * Callback for when an item is selected from the dropdown
   */
  onSelect: _propTypes["default"].func,

  /**
   * Aligns the dropdown to the right or left of the button opening the dropdown
   */
  alignRight: _propTypes["default"].bool,

  /**
   * Custom classname to apply to the Menu
   */
  menuClassName: _propTypes["default"].string,

  /**
   * Whether to activate using a long click
   */
  holdAffordance: _propTypes["default"].bool
};
