// @react/react-spectrum/Tour/js/Tour.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.STATUS_COMPLETE = exports.STATUS_CANCEL = void 0;

var _autobindDecorator = _interopRequireDefault(require("autobind-decorator"));

var _events = require("../../utils/events");

var _intl = require("../../utils/intl");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _class, _class2, _temp;

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var intlMessages = {
  "cs-CZ": _interopRequireDefault(require("./../intl/cs-CZ.json"))["default"],
  "da-DK": _interopRequireDefault(require("./../intl/da-DK.json"))["default"],
  "de-DE": _interopRequireDefault(require("./../intl/de-DE.json"))["default"],
  "en-US": _interopRequireDefault(require("./../intl/en-US.json"))["default"],
  "es-ES": _interopRequireDefault(require("./../intl/es-ES.json"))["default"],
  "fi-FI": _interopRequireDefault(require("./../intl/fi-FI.json"))["default"],
  "fr-FR": _interopRequireDefault(require("./../intl/fr-FR.json"))["default"],
  "hu-HU": _interopRequireDefault(require("./../intl/hu-HU.json"))["default"],
  "it-IT": _interopRequireDefault(require("./../intl/it-IT.json"))["default"],
  "ja-JP": _interopRequireDefault(require("./../intl/ja-JP.json"))["default"],
  "ko-KR": _interopRequireDefault(require("./../intl/ko-KR.json"))["default"],
  "nb-NO": _interopRequireDefault(require("./../intl/nb-NO.json"))["default"],
  "nl-NL": _interopRequireDefault(require("./../intl/nl-NL.json"))["default"],
  "pl-PL": _interopRequireDefault(require("./../intl/pl-PL.json"))["default"],
  "pt-BR": _interopRequireDefault(require("./../intl/pt-BR.json"))["default"],
  "ru-RU": _interopRequireDefault(require("./../intl/ru-RU.json"))["default"],
  "sv-SE": _interopRequireDefault(require("./../intl/sv-SE.json"))["default"],
  "tr-TR": _interopRequireDefault(require("./../intl/tr-TR.json"))["default"],
  "uk-UA": _interopRequireDefault(require("./../intl/uk-UA.json"))["default"],
  "zh-CN": _interopRequireDefault(require("./../intl/zh-CN.json"))["default"],
  "zh-TW": _interopRequireDefault(require("./../intl/zh-TW.json"))["default"]
};
var formatMessage = (0, _intl.messageFormatter)(intlMessages);
var STATUS_CANCEL = 'cancel';
exports.STATUS_CANCEL = STATUS_CANCEL;
var STATUS_COMPLETE = 'complete';
exports.STATUS_COMPLETE = STATUS_COMPLETE;

var Tour = (0, _autobindDecorator["default"])(_class = (_temp = _class2 =
/*#__PURE__*/
function (_React$Component) {
  _inherits(Tour, _React$Component);

  function Tour(props) {
    var _this;

    _classCallCheck(this, Tour);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(Tour).call(this, props));
    _this.state = {
      current: 0,
      total: 0,
      hidden: false
    };
    return _this;
  }

  _createClass(Tour, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      if (this.props.children) {
        this.setState({
          total: _react["default"].Children.count(this.props.children)
        });
      }
    }
  }, {
    key: "onConfirm",
    value: function onConfirm() {
      if (this.state.current === this.state.total - 1) {
        this.setState({
          hidden: true
        });

        if (this.props.onTourEnd) {
          this.props.onTourEnd(STATUS_COMPLETE);
        }
      } else {
        this.setState({
          current: this.state.current + 1
        });
      }
    }
  }, {
    key: "onCancel",
    value: function onCancel() {
      this.setState({
        current: 0,
        hidden: true
      });

      if (this.props.onTourEnd) {
        this.props.onTourEnd(STATUS_CANCEL);
      }
    }
  }, {
    key: "onHide",
    value: function onHide(event, currentChild) {
      if (this.props.clickOutsideAction === 'next') {
        (0, _events.chain)(this.onConfirm, currentChild.props.onConfirm)(event);
      } else if (this.props.clickOutsideAction === 'skip') {
        (0, _events.chain)(this.onCancel, currentChild.props.onCancel)(event);
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          children = _this$props.children,
          disableProgress = _this$props.disableProgress,
          clickOutsideAction = _this$props.clickOutsideAction;
      var _this$state = this.state,
          current = _this$state.current,
          total = _this$state.total,
          hidden = _this$state.hidden;

      if (!children || _react["default"].Children.count(children) <= this.state.current) {
        throw new Error('Tour should always have children!');
      }

      if (hidden) {
        return null;
      }

      var currentChild = _react["default"].Children.toArray(this.props.children)[current];

      return _react["default"].cloneElement(currentChild, {
        disableProgress: disableProgress,
        currentStep: current + 1,
        totalSteps: total,
        dismissible: clickOutsideAction === 'skip' || clickOutsideAction === 'next' ? true : undefined,
        onConfirm: (0, _events.chain)(this.onConfirm, currentChild.props.onConfirm),
        onCancel: (0, _events.chain)(this.onCancel, currentChild.props.onCancel),
        onHide: function onHide(event) {
          return _this2.onHide(event, currentChild);
        },
        confirmLabel: currentChild.props.confirmLabel || (current < total - 1 ? formatMessage('next') : formatMessage('done')),
        cancelLabel: currentChild.props.cancelLabel || (current < total - 1 ? formatMessage('skip') : null)
      });
    }
  }]);

  return Tour;
}(_react["default"].Component), _defineProperty(_class2, "propTypes", {
  /** Hide progress, the current step/total steps number in the
   *  top right of a coachmark, this is passed through */
  disableProgress: _propTypes["default"].bool,

  /** when not clicking on the coachmark itself, this determines the behavior.
   *  - 'next': a click off will behave the same as clicking 'next' on a coachmark
   *  - 'skip': a click off will skip the remainder of the tour
   *  - 'noop': a click off will do nothing
   **/
  clickOutsideAction: _propTypes["default"].oneOf(['next', 'skip', 'noop']),

  /**
   * Event when tour is finished. It's passed 'cancel' if the tour ended via the skip button
   * or other form of early dismissal. It's passed 'complete' if the tour is completed.
   */
  onTourEnd: _propTypes["default"].func
}), _defineProperty(_class2, "defaultProps", {
  disableProgress: false,
  clickOutsideAction: 'noop'
}), _temp)) || _class;

exports["default"] = Tour;
