// @react/react-spectrum/CoachMark/js/CoachMarkPopover.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _autobindDecorator = _interopRequireDefault(require("autobind-decorator"));

var _Button = _interopRequireDefault(require("../../Button"));

var _createId = _interopRequireDefault(require("../../utils/createId"));

var _filterDOMProps = _interopRequireDefault(require("../../utils/filterDOMProps"));

var _intl = require("../../utils/intl");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _FocusManager = require("../../utils/FocusManager");

require("../style/index.css");

var _class, _class2, _temp;

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var intlMessages = {
  "cs-CZ": _interopRequireDefault(require("./../intl/cs-CZ.json"))["default"],
  "da-DK": _interopRequireDefault(require("./../intl/da-DK.json"))["default"],
  "de-DE": _interopRequireDefault(require("./../intl/de-DE.json"))["default"],
  "en-US": _interopRequireDefault(require("./../intl/en-US.json"))["default"],
  "es-ES": _interopRequireDefault(require("./../intl/es-ES.json"))["default"],
  "fi-FI": _interopRequireDefault(require("./../intl/fi-FI.json"))["default"],
  "fr-FR": _interopRequireDefault(require("./../intl/fr-FR.json"))["default"],
  "hu-HU": _interopRequireDefault(require("./../intl/hu-HU.json"))["default"],
  "it-IT": _interopRequireDefault(require("./../intl/it-IT.json"))["default"],
  "ja-JP": _interopRequireDefault(require("./../intl/ja-JP.json"))["default"],
  "ko-KR": _interopRequireDefault(require("./../intl/ko-KR.json"))["default"],
  "nb-NO": _interopRequireDefault(require("./../intl/nb-NO.json"))["default"],
  "nl-NL": _interopRequireDefault(require("./../intl/nl-NL.json"))["default"],
  "pl-PL": _interopRequireDefault(require("./../intl/pl-PL.json"))["default"],
  "pt-BR": _interopRequireDefault(require("./../intl/pt-BR.json"))["default"],
  "ru-RU": _interopRequireDefault(require("./../intl/ru-RU.json"))["default"],
  "sv-SE": _interopRequireDefault(require("./../intl/sv-SE.json"))["default"],
  "tr-TR": _interopRequireDefault(require("./../intl/tr-TR.json"))["default"],
  "uk-UA": _interopRequireDefault(require("./../intl/uk-UA.json"))["default"],
  "zh-CN": _interopRequireDefault(require("./../intl/zh-CN.json"))["default"],
  "zh-TW": _interopRequireDefault(require("./../intl/zh-TW.json"))["default"]
};
var formatMessage = (0, _intl.messageFormatter)(intlMessages);

var CoachMarkPopover = (0, _autobindDecorator["default"])(_class = (_temp = _class2 =
/*#__PURE__*/
function (_React$Component) {
  _inherits(CoachMarkPopover, _React$Component);

  function CoachMarkPopover(props) {
    var _this;

    _classCallCheck(this, CoachMarkPopover);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(CoachMarkPopover).call(this, props));
    _this.coachmarkId = (0, _createId["default"])();
    return _this;
  }

  _createClass(CoachMarkPopover, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      var _this2 = this;

      if (this.props.trapFocus) {
        this._trapFocusTimeout = requestAnimationFrame(function () {
          if (_this2.coachmarkRef && !_this2.coachmarkRef.contains(document.activeElement)) {
            _this2.coachmarkRef.focus();
          }
        });
      }
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      if (this._trapFocusTimeout) {
        cancelAnimationFrame(this._trapFocusTimeout);
      }
    }
  }, {
    key: "onFocus",
    value: function onFocus(e) {
      if (this.props.onFocus) {
        this.props.onFocus(e);
      }

      if (this.props.trapFocus) {
        (0, _FocusManager.trapFocus)(this, e);
      }
    }
  }, {
    key: "onKeyDown",
    value: function onKeyDown(e) {
      if (this.props.onKeyDown) {
        this.props.onKeyDown(e); // Do nothing if stopPropagation has been called on event after onKeyDown prop executes.

        if (e.isPropagationStopped && e.isPropagationStopped()) {
          return;
        }
      }

      if (this.props.trapFocus) {
        (0, _FocusManager.trapFocus)(this, e);
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this3 = this;

      var _this$props = this.props,
          title = _this$props.title,
          confirmLabel = _this$props.confirmLabel,
          onConfirm = _this$props.onConfirm,
          cancelLabel = _this$props.cancelLabel,
          onCancel = _this$props.onCancel,
          disableProgress = _this$props.disableProgress,
          currentStep = _this$props.currentStep,
          totalSteps = _this$props.totalSteps,
          image = _this$props.image,
          children = _this$props.children,
          _this$props$id = _this$props.id,
          id = _this$props$id === void 0 ? this.coachmarkId : _this$props$id,
          tabIndex = _this$props.tabIndex,
          _this$props$imageAlt = _this$props.imageAlt,
          imageAlt = _this$props$imageAlt === void 0 ? '' : _this$props$imageAlt,
          autoFocus = _this$props.autoFocus,
          trapFocus = _this$props.trapFocus,
          otherProps = _objectWithoutProperties(_this$props, ["title", "confirmLabel", "onConfirm", "cancelLabel", "onCancel", "disableProgress", "currentStep", "totalSteps", "image", "children", "id", "tabIndex", "imageAlt", "autoFocus", "trapFocus"]);

      disableProgress = disableProgress || typeof currentStep !== 'number' || typeof totalSteps !== 'number'; // eslint-disable-next-line jsx-a11y/no-noninteractive-element-interactions

      return _react["default"].createElement("div", _extends({}, (0, _filterDOMProps["default"])(otherProps), {
        className: "spectrum-CoachMarkPopover",
        role: "dialog",
        "aria-labelledby": "".concat(id, "-header"),
        "aria-describedby": "".concat(id, "-content"),
        id: id,
        ref: function ref(c) {
          return _this3.coachmarkRef = c;
        },
        onFocus: this.onFocus,
        onKeyDown: this.onKeyDown,
        tabIndex: trapFocus && tabIndex == null ? 1 : tabIndex
      }), image && _react["default"].createElement("img", {
        src: image,
        alt: imageAlt,
        className: "spectrum-CoachMarkPopover-image"
      }), _react["default"].createElement("div", {
        className: "spectrum-CoachMarkPopover-header",
        id: "".concat(id, "-header")
      }, _react["default"].createElement("div", {
        className: "spectrum-CoachMarkPopover-title"
      }, title), !disableProgress && _react["default"].createElement("div", {
        className: "spectrum-CoachMarkPopover-step"
      }, formatMessage('steps', {
        currentStep: currentStep,
        totalSteps: totalSteps
      }))), _react["default"].createElement("div", {
        className: "spectrum-CoachMarkPopover-content",
        id: "".concat(id, "-content")
      }, children), _react["default"].createElement("div", {
        className: "spectrum-CoachMarkPopover-footer"
      }, cancelLabel && _react["default"].createElement(_Button["default"], {
        quiet: true,
        onClick: onCancel,
        autoFocus: !confirmLabel ? autoFocus : null
      }, cancelLabel), confirmLabel && _react["default"].createElement(_Button["default"], {
        variant: "primary",
        onClick: onConfirm,
        autoFocus: autoFocus
      }, confirmLabel)));
    }
  }]);

  return CoachMarkPopover;
}(_react["default"].Component), _defineProperty(_class2, "propTypes", {
  /** Title */
  title: _propTypes["default"].string.isRequired,

  /**
   * Text of the primary button.
   * Required if using Coachmarks on their own. Can't use `.required`
   * though because Tour can provide a default.
   **/
  confirmLabel: _propTypes["default"].string,

  /** Event Handler for the confirm button */
  onConfirm: _propTypes["default"].func,

  /** Text of the secondary button */
  cancelLabel: _propTypes["default"].string,

  /** Event Handler for the cancel button */
  onCancel: _propTypes["default"].func,

  /** Hide progress */
  disableProgress: _propTypes["default"].bool,

  /** Url to an image to display */
  image: _propTypes["default"].string,

  /** Text alternative for image */
  imageAlt: _propTypes["default"].string,

  /** Current step */
  currentStep: _propTypes["default"].number,

  /** Total amount of steps */
  totalSteps: _propTypes["default"].number,

  /** Whether to autofocus on mount */
  autoFocus: _propTypes["default"].bool,

  /** Whether to trapFocus within coachmark */
  trapFocus: _propTypes["default"].bool,

  /** tabIndex property to manage how coachmark popover is included in the keyboard navigation order */
  tabIndex: _propTypes["default"].number,

  /** String that will override the default id generated by the instance */
  id: _propTypes["default"].string
}), _defineProperty(_class2, "defaultProps", {
  autoFocus: true,
  imageAlt: '',
  trapFocus: true
}), _temp)) || _class;

exports["default"] = CoachMarkPopover;
