// @react/react-spectrum/CoachMark/js/CoachMark.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _autobindDecorator = _interopRequireDefault(require("autobind-decorator"));

var _closest = _interopRequireDefault(require("dom-helpers/query/closest"));

var _CoachMarkIndicator = _interopRequireDefault(require("./CoachMarkIndicator"));

var _CoachMarkPopover = _interopRequireDefault(require("./CoachMarkPopover"));

var _OverlayTrigger = _interopRequireDefault(require("../../OverlayTrigger/js/OverlayTrigger"));

var _Portal = _interopRequireDefault(require("react-overlays/lib/Portal"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _class, _class2, _temp;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

if (process.env.SCALE_MEDIUM && process.env.SCALE_LARGE) {
  require("../../spectrum-css/coachmark/index.css");

  require("../../spectrum-css/coachmark/index-diff.css");
} else if (process.env.SCALE_LARGE) {
  require("../../spectrum-css/coachmark/index-lg.css");
} else {
  require("../../spectrum-css/coachmark/index.css");
}

if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_LIGHT) require("../../spectrum-css/coachmark/multiStops/light.css");
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_DARK) require("../../spectrum-css/coachmark/multiStops/dark.css");
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_LIGHTEST) require("../../spectrum-css/coachmark/multiStops/lightest.css");
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_DARKEST) require("../../spectrum-css/coachmark/multiStops/darkest.css");

var CoachMark = (0, _autobindDecorator["default"])(_class = (_temp = _class2 =
/*#__PURE__*/
function (_React$Component) {
  _inherits(CoachMark, _React$Component);

  function CoachMark(props) {
    var _this;

    _classCallCheck(this, CoachMark);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(CoachMark).call(this, props));
    _this.state = {
      overlayContainer: null,
      indicatorPositioned: false
    };
    _this.shouldUpdatePosition = true;
    return _this;
  }

  _createClass(CoachMark, [{
    key: "getOverlayContainer",
    value: function getOverlayContainer() {
      if (this.node) {
        return (0, _closest["default"])(this.node, '.react-spectrum-provider') || document.body;
      }

      return null;
    }
  }, {
    key: "componentDidMount",
    value: function componentDidMount() {
      var overlayContainer = this.getOverlayContainer();

      if (overlayContainer) {
        this.setState({
          overlayContainer: overlayContainer
        });
      }
    }
  }, {
    key: "onHide",
    value: function onHide() {
      if (this.props.onHide) {
        this.props.onHide();
      }
    }
    /**
     * Because the indicator has to position itself based on its measured size
     * we'll already be a few renders in and the Overlay will be rendered
     * to the wrong position because the indicator isn't in place yet.
     * Once the indicator is in position, it'll emit this event. It'll also
     * emit it every following time, we want to for the Overlay trigger to
     * update the position only the once though, after that, the internals
     * of Overlay will take care of repositioning accurately.
     * This logic tells the state that the indicator has been positioned and we should
     * update to position of the Overlay.
     * There's also logic in componentDidUpdate to make sure it's only called once.
     **/

  }, {
    key: "onIndicatorPositioned",
    value: function onIndicatorPositioned() {
      if (!this.state.indicatorPositioned && this.shouldUpdatePosition) {
        this.setState({
          indicatorPositioned: true
        });
      }
    }
    /**
     * Once the indicator is positioned and we've run through a render
     * where we updated the Overlay, then we should no longer
     * need to force the Overlay to update its position.
     **/

  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate() {
      if (this.state.indicatorPositioned) {
        this.shouldUpdatePosition = false;
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          quiet = _this$props.quiet,
          selector = _this$props.selector,
          children = _this$props.children,
          dismissible = _this$props.dismissible,
          flip = _this$props.flip,
          placement = _this$props.placement,
          otherProps = _objectWithoutProperties(_this$props, ["quiet", "selector", "children", "dismissible", "flip", "placement"]);

      var overlayContainer = this.state.overlayContainer;
      return _react["default"].createElement("div", {
        ref: function ref(node) {
          return _this2.node = node;
        },
        style: {
          display: 'none'
        }
      }, " ", overlayContainer && _react["default"].createElement(_Portal["default"], {
        container: overlayContainer
      }, _react["default"].createElement(_OverlayTrigger["default"], {
        trigger: dismissible ? 'click' : undefined,
        show: dismissible ? undefined : true // {undefined} will force OverlayTrigger to default to {defaultShow}
        ,
        defaultShow: true,
        flip: flip,
        placement: placement,
        offset: CoachMark.INDICATOR_OFFSET,
        onHide: this.onHide,
        shouldUpdatePosition: this.state.indicatorPositioned && this.shouldUpdatePosition
      }, _react["default"].createElement(_CoachMarkIndicator["default"], {
        selector: selector,
        quiet: quiet,
        onPositioned: this.onIndicatorPositioned
      }), _react["default"].createElement(_CoachMarkPopover["default"], otherProps, children))));
    }
  }]);

  return CoachMark;
}(_react["default"].Component), _defineProperty(_class2, "propTypes", _objectSpread({
  /**
   * Used by overlay trigger
   */
  parentNode: _propTypes["default"].any,

  /**
   * If there isn't enough space for the coachmark,
   * should it flip across the axis to try and get more space
   **/
  flip: _propTypes["default"].bool,

  /**
   * Relative to the target element, where should the coachmark render
   */
  placement: _propTypes["default"].oneOf(['bottom', 'bottom left', 'bottom right', 'top', 'top left', 'top right', 'right', 'right bottom', 'right top', 'left', 'left bottom', 'left top']),

  /**
   * If someone clicks off the coachmark, then it will hide/dismiss
   */
  dismissible: _propTypes["default"].bool
}, _CoachMarkPopover["default"].propTypes, {}, _CoachMarkIndicator["default"].propTypes)), _defineProperty(_class2, "defaultProps", _objectSpread({
  flip: true,
  placement: 'right'
}, _CoachMarkPopover["default"].defaultProps, {}, _CoachMarkIndicator["default"].defaultProps)), _defineProperty(_class2, "INDICATOR_OFFSET", 6), _temp)) || _class;

exports["default"] = CoachMark;
