// @react/react-spectrum/utils/LiveRegionAnnouncer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LiveRegionMessage = exports.MessageBlock = exports.LiveRegion = exports["default"] = void 0;

var _convertUnsafeMethod = _interopRequireDefault(require("./convertUnsafeMethod"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireWildcard(require("react"));

var _reactDom = _interopRequireDefault(require("react-dom"));

require("./style/index.css");

var _VisuallyHidden = _interopRequireDefault(require("../VisuallyHidden"));

var _class, _class2, _temp;

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; if (obj != null) { var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/* Inspired by https://github.com/AlmeroSteyn/react-aria-live */
var liveRegionAnnouncer = null;
var node = null;
var clearAssertiveTimeoutId = null;
var clearPoliteTimeoutId = null;
var LIVEREGION_TIMEOUT_DELAY = 1000;

var LiveRegionAnnouncer =
/*#__PURE__*/
function (_Component) {
  _inherits(LiveRegionAnnouncer, _Component);

  function LiveRegionAnnouncer() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, LiveRegionAnnouncer);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(LiveRegionAnnouncer)).call.apply(_getPrototypeOf2, [this].concat(args)));

    _defineProperty(_assertThisInitialized(_this), "state", {
      assertiveMessage: '',
      politeMessage: ''
    });

    return _this;
  }

  _createClass(LiveRegionAnnouncer, [{
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      if (node) {
        document.body.removeChild(node);
      }

      liveRegionAnnouncer = null;
      node = null;
    }
  }, {
    key: "render",
    value: function render() {
      var _this$state = this.state,
          assertiveMessage = _this$state.assertiveMessage,
          politeMessage = _this$state.politeMessage;
      return _react["default"].createElement(LiveRegion, {
        assertiveMessage: assertiveMessage,
        politeMessage: politeMessage
      });
    }
  }], [{
    key: "getInstance",
    value: function getInstance(callback, context) {
      if (!liveRegionAnnouncer) {
        node = document.createElement('div');
        document.body.appendChild(node);

        _reactDom["default"].render(_react["default"].createElement(LiveRegionAnnouncer, {
          ref: function ref(l) {
            return liveRegionAnnouncer = l;
          }
        }), node, callback);
      } else if (callback) {
        callback();
      }

      return liveRegionAnnouncer;
    }
  }, {
    key: "destroyInstance",
    value: function destroyInstance(callback) {
      if (liveRegionAnnouncer) {
        liveRegionAnnouncer.setState({
          assertiveMessage: '',
          politeMessage: ''
        }, function () {
          var instanceNode = _reactDom["default"].findDOMNode(liveRegionAnnouncer);

          instanceNode && _reactDom["default"].unmountComponentAtNode(instanceNode.parentNode);

          if (callback) {
            // wait a frame for component to unmount
            setTimeout(callback, 50);
          }
        });
      }
    }
  }, {
    key: "announceAssertive",
    value: function announceAssertive(message) {
      var timeoutDuration = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : LIVEREGION_TIMEOUT_DELAY;

      if (clearAssertiveTimeoutId) {
        clearTimeout(clearAssertiveTimeoutId);
        clearAssertiveTimeoutId = null;
      }

      LiveRegionAnnouncer.getInstance(function () {
        return liveRegionAnnouncer.setState({
          assertiveMessage: message
        }, function () {
          if (message !== '') {
            clearAssertiveTimeoutId = setTimeout(function () {
              return LiveRegionAnnouncer.clearMessage('assertive');
            }, timeoutDuration);
          }
        });
      });
    }
  }, {
    key: "announcePolite",
    value: function announcePolite(message) {
      var timeoutDuration = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : LIVEREGION_TIMEOUT_DELAY;

      if (clearPoliteTimeoutId) {
        clearTimeout(clearPoliteTimeoutId);
        clearPoliteTimeoutId = null;
      }

      LiveRegionAnnouncer.getInstance(function () {
        return liveRegionAnnouncer.setState({
          politeMessage: message
        }, function () {
          if (message !== '') {
            clearPoliteTimeoutId = setTimeout(function () {
              return LiveRegionAnnouncer.clearMessage('polite');
            }, timeoutDuration);
          }
        });
      });
    }
  }, {
    key: "clearMessage",
    value: function clearMessage(politeness) {
      LiveRegionAnnouncer.getInstance(function () {
        if (!politeness) {
          LiveRegionAnnouncer.announceAssertive('');
          LiveRegionAnnouncer.announcePolite('');
        } else {
          liveRegionAnnouncer.setState(_defineProperty({}, politeness + 'Message', ''));
        }
      });
    }
  }]);

  return LiveRegionAnnouncer;
}(_react.Component);

exports["default"] = LiveRegionAnnouncer;

var LiveRegion = (0, _convertUnsafeMethod["default"])(_class = (_temp = _class2 =
/*#__PURE__*/
function (_Component2) {
  _inherits(LiveRegion, _Component2);

  function LiveRegion() {
    var _getPrototypeOf3;

    var _this2;

    _classCallCheck(this, LiveRegion);

    for (var _len2 = arguments.length, args = new Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
      args[_key2] = arguments[_key2];
    }

    _this2 = _possibleConstructorReturn(this, (_getPrototypeOf3 = _getPrototypeOf(LiveRegion)).call.apply(_getPrototypeOf3, [this].concat(args)));

    _defineProperty(_assertThisInitialized(_this2), "state", {
      assertiveMessage1: '',
      assertiveMessage2: '',
      politeMessage1: '',
      politeMessage2: ''
    });

    _defineProperty(_assertThisInitialized(_this2), "useAlternatePolite", false);

    _defineProperty(_assertThisInitialized(_this2), "useAlternateAssertive", false);

    return _this2;
  }

  _createClass(LiveRegion, [{
    key: "UNSAFE_componentWillReceiveProps",
    value: function UNSAFE_componentWillReceiveProps(nextProps) {
      var _this$props = this.props,
          oldAssertiveMessage = _this$props.assertiveMessage,
          oldPoliteMessage = _this$props.politeMessage;
      var assertiveMessage = nextProps.assertiveMessage,
          politeMessage = nextProps.politeMessage;

      if (oldPoliteMessage !== politeMessage) {
        this.setState({
          politeMessage1: this.useAlternatePolite ? '' : politeMessage,
          politeMessage2: this.useAlternatePolite ? politeMessage : ''
        });
        this.useAlternatePolite = !this.useAlternatePolite;
      }

      if (oldAssertiveMessage !== assertiveMessage) {
        this.setState({
          assertiveMessage1: this.useAlternateAssertive ? '' : assertiveMessage,
          assertiveMessage2: this.useAlternateAssertive ? assertiveMessage : ''
        });
        this.useAlternateAssertive = !this.useAlternateAssertive;
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this$state2 = this.state,
          assertiveMessage1 = _this$state2.assertiveMessage1,
          assertiveMessage2 = _this$state2.assertiveMessage2,
          politeMessage1 = _this$state2.politeMessage1,
          politeMessage2 = _this$state2.politeMessage2;
      return _react["default"].createElement(_VisuallyHidden["default"], {
        element: "div"
      }, _react["default"].createElement(MessageBlock, {
        "aria-live": "assertive",
        message: assertiveMessage1
      }), _react["default"].createElement(MessageBlock, {
        "aria-live": "assertive",
        message: assertiveMessage2
      }), _react["default"].createElement(MessageBlock, {
        "aria-live": "polite",
        message: politeMessage1
      }), _react["default"].createElement(MessageBlock, {
        "aria-live": "polite",
        message: politeMessage2
      }));
    }
  }]);

  return LiveRegion;
}(_react.Component), _defineProperty(_class2, "propTypes", {
  assertiveMessage: _propTypes["default"].string,
  politeMessage: _propTypes["default"].string
}), _temp)) || _class;

exports.LiveRegion = LiveRegion;

var MessageBlock = function MessageBlock(_ref) {
  var message = _ref.message,
      ariaLive = _ref['aria-live'];
  return _react["default"].createElement(_VisuallyHidden["default"], {
    element: "div",
    "aria-live": ariaLive,
    "aria-relevant": "additions",
    "aria-atomic": "true"
  }, message || '');
};

exports.MessageBlock = MessageBlock;
MessageBlock.propTypes = {
  message: _propTypes["default"].string,
  'aria-live': _propTypes["default"].string.isRequired
};

var LiveRegionMessage =
/*#__PURE__*/
function (_Component3) {
  _inherits(LiveRegionMessage, _Component3);

  function LiveRegionMessage() {
    _classCallCheck(this, LiveRegionMessage);

    return _possibleConstructorReturn(this, _getPrototypeOf(LiveRegionMessage).apply(this, arguments));
  }

  _createClass(LiveRegionMessage, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this.announce();
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      if (this.props.message !== prevProps.message) {
        this.announce();
      }
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      if (this.props.clearOnUnmount) {
        LiveRegionAnnouncer.clearMessage();
      }
    }
  }, {
    key: "announce",
    value: function announce() {
      var _this$props2 = this.props,
          message = _this$props2.message,
          ariaLive = _this$props2['aria-live'];

      if (ariaLive === 'assertive') {
        LiveRegionAnnouncer.announceAssertive(message || '');
      }

      if (ariaLive === 'polite') {
        LiveRegionAnnouncer.announcePolite(message || '');
      }
    }
  }, {
    key: "render",
    value: function render() {
      return null;
    }
  }]);

  return LiveRegionMessage;
}(_react.Component);

exports.LiveRegionMessage = LiveRegionMessage;

_defineProperty(LiveRegionMessage, "propTypes", {
  message: _propTypes["default"].string.isRequired,
  'aria-live': _propTypes["default"].string.isRequired,
  clearOnUnmount: _propTypes["default"].bool
});
