// @react/react-spectrum/Search/js/Search.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _autobindDecorator = _interopRequireDefault(require("autobind-decorator"));

var _Button = _interopRequireDefault(require("../../Button"));

var _classnames = _interopRequireDefault(require("classnames"));

var _icon = require("../../utils/icon");

var _convertUnsafeMethod = _interopRequireDefault(require("../../utils/convertUnsafeMethod"));

var _createId = _interopRequireDefault(require("../../utils/createId"));

var _CrossSmall = _interopRequireDefault(require("../../Icon/core/CrossSmall"));

var _Magnifier = _interopRequireDefault(require("../../Icon/core/Magnifier"));

var _intl = require("../../utils/intl");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireWildcard(require("react"));

var _Textfield = _interopRequireDefault(require("../../Textfield"));

var _class, _class2, _temp;

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; if (obj != null) { var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var intlMessages = {
  "cs-CZ": _interopRequireDefault(require("./../intl/cs-CZ.json"))["default"],
  "da-DK": _interopRequireDefault(require("./../intl/da-DK.json"))["default"],
  "de-DE": _interopRequireDefault(require("./../intl/de-DE.json"))["default"],
  "en-US": _interopRequireDefault(require("./../intl/en-US.json"))["default"],
  "es-ES": _interopRequireDefault(require("./../intl/es-ES.json"))["default"],
  "fi-FI": _interopRequireDefault(require("./../intl/fi-FI.json"))["default"],
  "fr-FR": _interopRequireDefault(require("./../intl/fr-FR.json"))["default"],
  "hu-HU": _interopRequireDefault(require("./../intl/hu-HU.json"))["default"],
  "it-IT": _interopRequireDefault(require("./../intl/it-IT.json"))["default"],
  "ja-JP": _interopRequireDefault(require("./../intl/ja-JP.json"))["default"],
  "ko-KR": _interopRequireDefault(require("./../intl/ko-KR.json"))["default"],
  "nb-NO": _interopRequireDefault(require("./../intl/nb-NO.json"))["default"],
  "nl-NL": _interopRequireDefault(require("./../intl/nl-NL.json"))["default"],
  "pl-PL": _interopRequireDefault(require("./../intl/pl-PL.json"))["default"],
  "pt-BR": _interopRequireDefault(require("./../intl/pt-BR.json"))["default"],
  "ru-RU": _interopRequireDefault(require("./../intl/ru-RU.json"))["default"],
  "sv-SE": _interopRequireDefault(require("./../intl/sv-SE.json"))["default"],
  "tr-TR": _interopRequireDefault(require("./../intl/tr-TR.json"))["default"],
  "uk-UA": _interopRequireDefault(require("./../intl/uk-UA.json"))["default"],
  "zh-CN": _interopRequireDefault(require("./../intl/zh-CN.json"))["default"],
  "zh-TW": _interopRequireDefault(require("./../intl/zh-TW.json"))["default"]
};

if (process.env.SCALE_MEDIUM && process.env.SCALE_LARGE) {
  require("../../spectrum-css/search/index.css");

  require("../../spectrum-css/search/index-diff.css");
} else if (process.env.SCALE_LARGE) {
  require("../../spectrum-css/search/index-lg.css");
} else {
  require("../../spectrum-css/search/index.css");
}

if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_LIGHT) require("../../spectrum-css/search/multiStops/light.css");
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_DARK) require("../../spectrum-css/search/multiStops/dark.css");
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_LIGHTEST) require("../../spectrum-css/search/multiStops/lightest.css");
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_DARKEST) require("../../spectrum-css/search/multiStops/darkest.css");
var formatMessage = (0, _intl.messageFormatter)(intlMessages);

var normalizeValue = function normalizeValue(value) {
  return value || '';
};

var Search = (0, _convertUnsafeMethod["default"])(_class = (0, _autobindDecorator["default"])(_class = (_temp = _class2 =
/*#__PURE__*/
function (_Component) {
  _inherits(Search, _Component);

  function Search(props) {
    var _this;

    _classCallCheck(this, Search);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(Search).call(this, props));
    _this.searchId = (0, _createId["default"])();
    var value = props.value,
        defaultValue = props.defaultValue;
    _this.state = {
      value: normalizeValue(value || defaultValue)
    };
    return _this;
  }

  _createClass(Search, [{
    key: "UNSAFE_componentWillReceiveProps",
    value: function UNSAFE_componentWillReceiveProps(props) {
      if (normalizeValue(props.value) !== this.state.value) {
        this.setState({
          value: props.value
        });
      }
    }
  }, {
    key: "handleTextKeyDown",
    value: function handleTextKeyDown(e) {
      var _this$props = this.props,
          onSubmit = _this$props.onSubmit,
          onKeyDown = _this$props.onKeyDown,
          disabled = _this$props.disabled;
      var value = this.state.value;
      var key = e.which;

      if (key === 13 || key === 27) {
        e.preventDefault();
      }

      if (disabled) {
        return;
      }

      if (key === 13) {
        onSubmit(value);
      }

      if (key === 27) {
        this.handleTextChange('', e, 'escapeKey');
      }

      if (onKeyDown) {
        onKeyDown(e);
      }
    }
  }, {
    key: "handleTextChange",
    value: function handleTextChange(value, e) {
      var from = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 'input';
      var _this$props2 = this.props,
          onChange = _this$props2.onChange,
          disabled = _this$props2.disabled;

      if (disabled || value === this.state.value) {
        return;
      }

      if (!('value' in this.props)) {
        this.setState({
          value: value
        });
      }

      onChange(value, e, {
        from: from
      });
    }
  }, {
    key: "handleClearButtonClick",
    value: function handleClearButtonClick(e) {
      this.handleTextChange('', e, 'clearButton'); // restore focus to the searchbox

      if (this.searchbox) {
        this.searchbox.focus();
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props3 = this.props,
          disabled = _this$props3.disabled,
          className = _this$props3.className,
          icon = _this$props3.icon,
          _this$props3$id = _this$props3.id,
          id = _this$props3$id === void 0 ? this.searchId : _this$props3$id,
          otherProps = _objectWithoutProperties(_this$props3, ["disabled", "className", "icon", "id"]);

      var value = this.state.value;
      return _react["default"].createElement("div", {
        className: (0, _classnames["default"])('spectrum-Search', {
          'is-disabled': disabled
        }, className)
      }, _react["default"].createElement(_Textfield["default"], _extends({
        type: "search",
        ref: function ref(s) {
          return _this2.searchbox = s;
        },
        className: "spectrum-Search-input",
        value: value,
        disabled: disabled,
        id: id
      }, otherProps, {
        onKeyDown: this.handleTextKeyDown,
        onChange: this.handleTextChange
      })), (0, _icon.cloneIcon)(icon, {
        className: 'spectrum-Search-icon',
        size: 'S'
      }), value !== '' && _react["default"].createElement(_Button["default"], {
        "aria-label": formatMessage('Clear search'),
        "aria-labelledby": "".concat(id, " ").concat(id, "-clear-button"),
        variant: "clear",
        icon: _react["default"].createElement(_CrossSmall["default"], null),
        id: "".concat(id, "-clear-button"),
        disabled: disabled,
        onClick: this.handleClearButtonClick
      }));
    }
  }]);

  return Search;
}(_react.Component), _defineProperty(_class2, "propTypes", {
  /** Sets the search field value (controlled). */
  value: _propTypes["default"].string,

  /** Sets the default search field value (uncontrolled). */
  defaultValue: _propTypes["default"].string,

  /** Sets the icon displayed in the search field. Magnifier icon by default. */
  icon: _propTypes["default"].node
}), _defineProperty(_class2, "defaultProps", {
  icon: _react["default"].createElement(_Magnifier["default"], null),
  onChange: function onChange() {},
  onSubmit: function onSubmit() {}
}), _temp)) || _class) || _class;

exports["default"] = Search;
Search.displayName = 'Search';
